{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/ydmatsd/funbgcex-colab2/blob/main/funbgcex-colab/FunBGCeX.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "<img src=\"https://github.com/ydmatsd/funbgcex-colab/blob/main/funbgcex-colab/logo.png?raw=true\" height=\"60\" align=\"top\">\n",
        "\n",
        "##<b>FunBGCeX (Fungal Biosynthetic Gene Cluster eXtractor) ‚Äì Google Colab Version</b>\n",
        "[**FunBGCeX**](https://github.com/ydmatsd/funbgcex) is a genome mining tool for fungal natural product discovery. It detects biosynthetic proteins based on our manually curated database of fungal natural product biosynthetic gene clusters (BGCs), [FunBGCs](http://staffweb1.cityu.edu.hk/ymatsuda/funbgcs/funbgcs.html). FunBGCeX can detect all possible BGCs from given fungal genomes and selectively extract BGCs according to the user's various needs. Currently available methods include:\n",
        "* Extraction of BGCs encoding a homologue of a protein of the user's interest\n",
        "* Extraction of BGCs harboring a putative self-resistance gene\n",
        "* Extraction of BGCs lacking well-defined core biosynthetic enzyme genes (e.g., PKSs, NRPSs, and terpene synthases) ‚Äî still under development\n",
        "\n",
        "The extracted BGCs are provided in the GenBank format and can also be visualized using a web browser (You can access a sample output [here](http://staffweb1.cityu.edu.hk/ymatsuda/funbgcex/sample2/results.html)).\n",
        "### Citation\n",
        "[Tang, J., Matsuda, Y. *Nat. Commun.* **15**, 4312 (2024)](https://www.nature.com/articles/s41467-024-48771-7)\n",
        "\n",
        "-----------"
      ],
      "metadata": {
        "id": "K2KgE-a8bg6x"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title ## <b>Provide Input Information for BGC Extraction</b>\n",
        "#@markdown ### Fill in the required settings below, then <b>ONLY</b> run this cell.\n",
        "#@markdown Depending on your chosen input and extraction modes, you may need to:\n",
        "#@markdown - Upload GenBank file(s)\n",
        "#@markdown - Specify a folder path (on Google Drive or Colab local)\n",
        "#@markdown - Upload a FASTA file (for `target` mode)\n",
        "#@markdown - Select a Pfam domain (for `pfam` mode)\n",
        "\n",
        "import os\n",
        "import shutil\n",
        "import time\n",
        "from google.colab import files, drive\n",
        "from IPython.display import display, clear_output\n",
        "import ipywidgets as widgets\n",
        "\n",
        "# ------------------------\n",
        "# Input settings\n",
        "# ------------------------\n",
        "#@markdown **Choose how to provide input data**\n",
        "InputMode = \"upload\" #@param [\"upload\", \"path\"]\n",
        "#@markdown  - `upload`: Upload GenBank file(s). <b>The GenBank files need to have CDS features with a translation qualifier.</b>\n",
        "#@markdown  - `path`: Specify an existing folder path (Google Drive or Colab local). <b>Ensure that no non-GenBank files are placed in this folder.</b>\n",
        "\n",
        "#@markdown **Only for `path` as InputMode**\n",
        "\n",
        "gbk_dir = \"/content/gbk_dir\" #@param {type:\"string\"}\n",
        "#@markdown - Provide the path of the folder containing GenBank files (e.g. `/content/my_data`).\n",
        "UseDrive = False #@param {type:\"boolean\"}\n",
        "#@markdown - Tick here if you wish to use Google Drive (the path to the directory will be `/content/drive/MyDrive/[your directory name]`).\n",
        "\n",
        "#@markdown **Choose a mode for BGC extraction**\n",
        "mode = \"all\" #@param [\"all\", \"target\", \"pfam\", \"ripps\", \"sre\", \"human\"]\n",
        "#@markdown  - `all`: Extract all possible BGCs.\n",
        "#@markdown  - `target`: Extract BGCs that encode a homologue of protein(s) of your interest (a FASTA file is required).\n",
        "#@markdown  - `pfam`: Extract BGCs that encode a protein with a protein family/domain (Pfam) of your interest.\n",
        "#@markdown  - `ripps`: Extract BGCs that encode a possible precursor peptide for RiPP biosynthesis.\n",
        "#@markdown  - `sre`: Extract BGCs with a possible self-resistance gene.\n",
        "#@markdown  - `human`: Extract BGCs that encode a homologue of human disease-related proteins.\n",
        "\n",
        "# ------------------------\n",
        "# Global variables\n",
        "# ------------------------\n",
        "fasta_path = None\n",
        "pfam = None\n",
        "\n",
        "# ------------------------\n",
        "# Functions\n",
        "# ------------------------\n",
        "def proceed_to_genbank_upload():\n",
        "    \"\"\"Handle GenBank upload / folder selection\"\"\"\n",
        "    global gbk_dir\n",
        "\n",
        "    clear_output()\n",
        "    if InputMode == \"upload\":\n",
        "        timestamp = int(time.time())\n",
        "        gbk_dir = f\"/content/gbk_upload_{timestamp}\"\n",
        "        os.makedirs(gbk_dir, exist_ok=True)\n",
        "        print(\"üìÇ Please upload your GenBank file(s)\")\n",
        "        uploaded = files.upload()\n",
        "        if not uploaded:\n",
        "            print(\"‚ö†Ô∏è No GenBank file uploaded.\")\n",
        "            return\n",
        "        for fname in uploaded.keys():\n",
        "            dest_path = os.path.join(gbk_dir, fname)\n",
        "            shutil.move(fname, dest_path)\n",
        "            print(f\"‚úÖ GenBank file saved to {dest_path}\")\n",
        "\n",
        "    elif InputMode == \"path\":\n",
        "        if UseDrive:\n",
        "            drive.mount('/content/drive')\n",
        "        print(f\"üìÅ Using existing folder: {gbk_dir}\")\n",
        "\n",
        "def upload_fasta():\n",
        "    \"\"\"Upload a single FASTA file\"\"\"\n",
        "    global fasta_path\n",
        "    fasta_dir = \"/content/fasta_input\"\n",
        "    os.makedirs(fasta_dir, exist_ok=True)\n",
        "    print(\"üìÇ Please upload your FASTA file (single file).\")\n",
        "    uploaded = files.upload()\n",
        "    if not uploaded:\n",
        "        print(\"‚ö†Ô∏è No FASTA file uploaded.\")\n",
        "        return\n",
        "    for fname in uploaded.keys():\n",
        "        dest_path = os.path.join(fasta_dir, fname)\n",
        "        shutil.move(fname, dest_path)\n",
        "        fasta_path = dest_path\n",
        "        print(f\"‚úÖ FASTA file saved to: {fasta_path}\")\n",
        "\n",
        "def select_pfam():\n",
        "    \"\"\"Select a Pfam domain\"\"\"\n",
        "    global pfam\n",
        "\n",
        "    pfam_list = ['2-Hacid_dh', '2-Hacid_dh_C', '2OG-FeII_Oxy', '2OG-FeII_Oxy_2', '2OG-FeII_Oxy_3', '2OG-Fe_Oxy_2', '3Beta_HSD', '3HCDH', '3HCDH_N', '4HBT', '4HBT_2', '4HBT_3', 'AA-adenyl-dom', 'AAA_16', 'AAA_19', 'AAA_21', 'AAA_22', 'AAA_23', 'AAA_25', 'AAA_29', 'AAA_30', 'AAA_33', 'AAT', 'AATase', 'AA_kinase', 'AA_permease', 'AA_permease_2', 'ABA3', 'ABC1', 'ABC2_membrane', 'ABC2_membrane_3', 'ABC2_membrane_7', 'ABC_ATPase', 'ABC_membrane', 'ABC_tran', 'ABC_trans_N', 'ABM', 'ACAS_N', 'ACC_central', 'ACPS', 'ACP_PCP', 'ACP_syn_III_C', 'ACT', 'ACT_7', 'ADC', 'ADH_N', 'ADH_N_2', 'ADH_zinc_N', 'ADH_zinc_N_2', 'ADP_ribosyl_GH', 'ADSL_C', 'AIG2_2', 'APH', 'ATEG_03636', 'ATP-grasp', 'ATP-grasp_4', 'ATP-synt_ab', 'ATP-synt_ab_N', 'ATP11', 'ATPgrasp_N', 'ATPgrasp_Ter', 'AXE1', 'Aa_trans', 'Abhydrolase_1', 'Abhydrolase_2', 'Abhydrolase_3', 'Abhydrolase_6', 'Acetyltransf_1', 'Acetyltransf_3', 'Acetyltransf_7', 'Acetyltransf_8', 'Aconitase', 'Aconitase_C', 'Acyl-CoA_dh_1', 'Acyl-CoA_dh_2', 'Acyl-CoA_dh_M', 'Acyl-CoA_dh_N', 'Acyl_transf_1', 'Acyl_transf_3', 'AflR', 'Aldedh', 'Aldo_ket_red', 'Aldolase_II', 'Aldose_epim', 'Alpha-amylase', 'Amidase', 'Amidohydro_1', 'Amidohydro_2', 'Amidohydro_3', 'Amino_oxidase', 'Aminotran_1_2', 'Aminotran_3', 'Aminotran_4', 'Aminotran_5', 'Ank', 'AnkA', 'AnkG', 'Ank_2', 'Ank_3', 'Ank_4', 'Ank_5', 'Anth_synt_I_N', 'Anthrone_oxy', 'ApbA', 'ApbA_C', 'Arginosuc_synth', 'AsR5', 'AsR6', 'Asn_synthase', 'Asp', 'Asp_Glu_race', 'Asp_aminotransf', 'AstC', 'AtaL', 'AtnE', 'BAAT_C', 'BBE', 'BD-FAE', 'Bac_luciferase', 'Beta-lactamase', 'Beta_elim_lyase', 'BfL6', 'BfL8', 'Biotin_carb_C', 'Biotin_carb_N', 'Biotin_lipoyl', 'CBM-like', 'CBS', 'CMAS', 'CMD', 'COesterase', 'CPSase_L_D2', 'CPSase_L_D3', 'CalD', 'Caps_synth', 'Carboxyl_trans', 'Carn_acyltransf', 'Catalase-rel', 'Catalase', 'CghF', 'Chal_sti_synt_C', 'Chal_sti_synt_N', 'Chlorophyllase2', 'Chorismate_bind', 'Citrate_synt', 'Clz3', 'Clz7', 'CoA_transf_3', 'Cofilin_ADF', 'Condensation', 'CosA', 'Cp1D', 'CpaE', 'CtvD', 'Cu-oxidase', 'Cu-oxidase_2', 'Cu-oxidase_3', 'Cu_amine_oxid', 'Cupin_2', 'Cys_Met_Meta_PP', 'Cyt-b5', 'Cytochrom_B561', 'DAHP_synth_1', 'DAO', 'DDE_1', 'DDE_3', 'DHDPS', 'DIOX_N', 'DIT1_PvcA', 'DJ-1_PfpI', 'DLH', 'DSBA', 'DUF1295', 'DUF1479', 'DUF1729', 'DUF1749', 'DUF1752', 'DUF1772', 'DUF1776', 'DUF1857', 'DUF1929', 'DUF1965', 'DUF1993', 'DUF2306', 'DUF2334', 'DUF2470', 'DUF2784', 'DUF3176', 'DUF3237', 'DUF3425', 'DUF3471', 'DUF3533', 'DUF4149', 'DUF4185', 'DUF4188', 'DUF4267', 'DUF4385', 'DUF4413', 'DUF4451', 'DUF5519', 'DUF6534', 'DUF6570', 'DUF659', 'DUF87', 'Dabb', 'Dala_Dala_lig_C', 'Dimer_Tnp_hAT', 'Dimerisation', 'Dimerisation2', 'Dynamin_N', 'Dyp_perox', 'ECH_1', 'ECH_2', 'EHN', 'ERG4_ERG24', 'EamA', 'EcKL', 'EcdJ', 'EfHP', 'Egh16-like', 'ElcH', 'Epimerase', 'EqxR', 'Esterase', 'EthD', 'F420_oxidored', 'F5_F8_type_C', 'FAA_hydrolase', 'FAD_binding_1', 'FAD_binding_2', 'FAD_binding_3', 'FAD_binding_4', 'FAD_binding_6', 'FAD_oxidored', 'FAE1_CUT1_RppA', 'FAS_I_H', 'FAS_N', 'FAS_meander', 'FA_hydroxylase', 'FMN_dh', 'FMN_red', 'FMO-like', 'FPSE_09189', 'FR47', 'FSH1', 'Fas_alpha_ACP', 'Fasciclin', 'Fe-ADH', 'Fe-ADH_2', 'Ferric_reduct', 'FhuF', 'Flavodoxin_1', 'Flavodoxin_2', 'Fn3-like', 'FncF', 'FolB', 'FrsA-like', 'FrzJ', 'Fsa2', 'FscB', 'FtsJ', 'FtsK_SpoIIIE', 'Fungal_trans', 'Fungal_trans_2', 'GATase', 'GATase_6', 'GATase_7', 'GBBH-like_N', 'GDP_Man_Dehyd', 'GFO_IDH_MocA', 'GFO_IDH_MocA_C', 'GGPS', 'GH3', 'GH43_C2', 'GIDA', 'GMC_oxred_C', 'GMC_oxred_N', 'GO-like_E_set', 'GST_C', 'GST_C_2', 'GST_C_3', 'GST_C_5', 'GST_N', 'GST_N_2', 'GST_N_3', 'GST_N_4', 'GTP_cyclohydroI', 'G_glu_transpept', 'GliK', 'GliN', 'Gln-synt_C', 'Glu_dehyd_C', 'Gly_transf_sug', 'Glyco_hydro_11', 'Glyco_hydro_3', 'Glyco_hydro_31', 'Glyco_hydro_3_C', 'Glyco_hydro_43', 'Glyco_hydro_cc', 'Glyco_tran_28_C', 'Glyco_transf_28', 'Glyoxal_oxid_N', 'Glyoxalase', 'Glyoxalase_3', 'Glyoxalase_4', 'Glyoxalase_6', 'Gp_dh_C', 'Gp_dh_N', 'Gpr1_Fun34_YaaH', 'GsfR1', 'HAD_2', 'HD', 'HI0933_like', 'HLH', 'HMG-CoA_red', 'HMGL-like', 'HMG_CoA_synt_C', 'HMG_CoA_synt_N', 'HPIH', 'HSP20', 'HTH_51', 'HTH_Tnp_Tc5', 'HTH_psq', 'Haem_degrading', 'HbpS-like', 'Helitron_like_N', 'Helo_like_N', 'Hemerythrin', 'Hemocyanin_C', 'Hemocyanin_M', 'Hexapep', 'HimD', 'Histone', 'Histone_H2A_C', 'Homeobox_KN', 'Homeodomain', 'Homoserine_dh', 'HpcH_HpaI', 'Hydrolase', 'Hydrolase_4', 'Hydrolase_like', 'ICMT', 'IDO', 'ILVD_EDD', 'IMPDH', 'IPPT', 'IPT', 'Iso_dh', 'Isochorismatase', 'IucA_IucC', 'K_oxygenase', 'Kdo', 'Kelch_1', 'Kelch_2', 'Kelch_3', 'Kelch_4', 'Kelch_6', 'LCM', 'LIP', 'Lactamase_B', 'Lactamase_B_2', 'Lactamase_B_3', 'LcsT', 'LigB', 'Lipocalin_5', 'Lipocalin_9', 'LolU', 'LtmS', 'Lumazine_bd_2', 'Lyase_1', 'Lyase_aromatic', 'Lycopene_cycl', 'Lys_Orn_oxgnase', 'Lysine_decarbox', 'M20_dimer', 'MARVEL', 'MBOAT_2', 'MFS_1', 'MFS_1_like', 'MFS_2', 'MFS_4', 'MFS_5', 'MMR_HSR1', 'MPAB_Lcp_cat', 'MTS', 'MULE', 'MafB19-deam', 'MaoC_dehydrat_N', 'MaoC_dehydratas', 'MdpA', 'Melibiase_2', 'Melibiase_C', 'MetW', 'Methyltransf_10', 'Methyltransf_11', 'Methyltransf_12', 'Methyltransf_16', 'Methyltransf_2', 'Methyltransf_21', 'Methyltransf_23', 'Methyltransf_24', 'Methyltransf_25', 'Methyltransf_28', 'Methyltransf_3', 'Methyltransf_31', 'Methyltransf_32', 'Methyltransf_33', 'Methyltransf_4', 'Methyltransf_7', 'Methyltransf_9', 'Mito_carr', 'MmgE_PrpD', 'MmgE_PrpD_C', 'MscL', 'Myb_DNA-bind_6', 'Myb_DNA-bind_7', 'Myb_DNA-binding', 'NACHT', 'NAD_binding_1', 'NAD_binding_10', 'NAD_binding_11', 'NAD_binding_2', 'NAD_binding_3', 'NAD_binding_4', 'NAD_binding_6', 'NAD_binding_8', 'NAD_binding_9', 'NMO', 'NRPS_C', 'NRPS_CT', 'NRPS_E', 'NRPS_term_dom', 'Nitroreductase', 'NmrA', 'NnrU', 'NodS', 'NtCtMGAM_N', 'NvfI', 'O-FucT', 'OATP', 'OPT', 'OphC', 'Orn_Arg_deC_N', 'Orn_DAP_Arg_deC', 'OtaR1', 'Oxidored_FMN', 'Oxysterol_BP', 'P5CR_dimer', 'PAP2_3', 'PBP', 'PCMT', 'PDR_CDR', 'PEMT', 'PEP_mutase', 'PHM7_cyt', 'PIF1', 'PIGO_PIGG', 'PKS_AT', 'PKS_KS', 'PKS_NRPS_MT', 'PK_Tyr_Ser-Thr', 'PPR_2', 'PSDC', 'PS_Dcarbxylase', 'PTR2', 'P_proprotein', 'PatF', 'PatJ', 'Patched', 'PaxC', 'PbcA', 'Pec_lyase', 'PenF', 'Peptidase_M19', 'Peptidase_M20', 'Peptidase_M24', 'Peptidase_M3', 'Peptidase_S10', 'Peptidase_S15', 'Peptidase_S28', 'Peptidase_S41', 'Peptidase_S8', 'Peptidase_S9', 'Peptidase_S9_N', 'Peroxidase_2', 'PfB', 'Phe_hydrox_dim', 'PhomB', 'Phosphodiest', 'PhyH', 'Pkinase', 'Polyketide_cyc2', 'Polysacc_deac_1', 'Polysacc_synt_2', 'Prefoldin', 'Prenyltrans', 'Pribosyltran', 'Pro_CA', 'Pro_isomerase', 'Proteasome', 'PrpF', 'PydY', 'Pyr4', 'Pyr_redox', 'Pyr_redox_2', 'Pyr_redox_3', 'Pyridoxal_deC', 'Questin_oxidase', 'RSN1_7TM', 'RSN1_TM', 'RTA1', 'RmlD_sub_bind', 'RsgA_GTPase', 'Rua1_C', 'SAT2', 'SET', 'SGL', 'SMC_N', 'SQHop_cyclase_C', 'SQHop_cyclase_N', 'SQS_PSY', 'Sacchrp_dh_NADP', 'SalTPS', 'Say1_Mug180', 'Scytalone_dh', 'Sdh_cyt', 'SdnG', 'SdnJ', 'Semialdhyde_dh', 'Shikimate_DH', 'SmbD', 'SnoaL', 'SnoaL_2', 'SnoaL_4', 'Spd14', 'SrdD', 'Steroid_dh', 'Sterol-sensing', 'Sterol_MT_C', 'Str_synth', 'Sugar_tr', 'Sulfotransfer_4', 'T2SSE', 'TFIIE_alpha', 'TIM', 'TPMT', 'TPP_enzyme_C', 'TPP_enzyme_M', 'TPP_enzyme_N', 'TPR_2', 'TP_methylase', 'TRAM1', 'TRAM_LAG1_CLN8', 'TRI12', 'TRI5', 'TRI9', 'TTL', 'Tannase', 'TauD', 'Tcf25', 'TdiC', 'Terpene_syn_C_2', 'Terpene_synth_C', 'Thi4', 'Thioesterase', 'Thiolase_N', 'Tmemb_14', 'TqaK', 'Transferase', 'Transketolase_C', 'Transketolase_N', 'Transp_cyt_pur', 'Tri14', 'Tri3', 'Tri6', 'Trp_DMAT', 'Trp_halogenase', 'Trt14', 'Trt5', 'Trypsin', 'Trypsin_2', 'TwmC', 'Tyrosinase', 'UDPGT', 'UNC-93', 'UbiA_PT', 'UbiA_TC', 'Ubie_methyltran', 'UstYa', 'VicA', 'VicYx', 'VidS', 'VniA', 'VrtF', 'WD40', 'XenC', 'XenD', 'YCII', 'Zn_clus', 'adh_short', 'adh_short_C2', 'bZIP_1', 'dCMP_cyt_deam_1', 'fPKS_DH', 'fPKS_ER', 'fPKS_KR', 'fPKS_PT', 'fPKS_R', 'fPKS_TH', 'fn3_3', 'p450', 'peroxidase', 'tRNA-synt_1c', 'tRNA-synt_1c_C', 'tRNA-synt_2', 'tRNA-synt_2c', 'tRNA-synt_2d', 'tRNA_anti-codon', 'zf-C2H2', 'zf-C6H2', 'zf-H2C2_2', 'zf_CCCH_4']\n",
        "\n",
        "    pfam_widget = widgets.Combobox(\n",
        "        placeholder='Type to search Pfam...',\n",
        "        options=pfam_list,\n",
        "        description='Pfam:',\n",
        "        ensure_option=True,\n",
        "        disabled=False\n",
        "    )\n",
        "    confirm_button = widgets.Button(description=\"Confirm Pfam\")\n",
        "\n",
        "    def on_confirm_clicked(b):\n",
        "        global pfam\n",
        "        pfam = pfam_widget.value\n",
        "        if pfam.strip() == \"\":\n",
        "            print(\"‚ö†Ô∏è Please select a Pfam domain before confirming.\")\n",
        "            return\n",
        "        clear_output()\n",
        "        print(f\"‚úÖ Pfam selected: {pfam}\")\n",
        "        proceed_to_genbank_upload()\n",
        "\n",
        "    confirm_button.on_click(on_confirm_clicked)\n",
        "    display(pfam_widget, confirm_button)\n",
        "\n",
        "# ------------------------\n",
        "# Main logic\n",
        "# ------------------------\n",
        "if mode == \"target\":\n",
        "    upload_fasta()\n",
        "    proceed_to_genbank_upload()\n",
        "elif mode == \"pfam\":\n",
        "    select_pfam()\n",
        "else:\n",
        "    proceed_to_genbank_upload()\n",
        "\n",
        "#Output directory\n",
        "results_dir = \"/content/result\""
      ],
      "metadata": {
        "cellView": "form",
        "id": "RVxHAqHjCg2n"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##After completing the first step, move to the next cell, then go to `Runtime` -> `Run after`. Adjust the optional settings below if needed.\n"
      ],
      "metadata": {
        "id": "IucgAcj1Y1lj"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Optional settings\n",
        "gap_allowed = 2500 #@param {type:\"integer\"}\n",
        "#@markdown - Maximum allowed gap (in bp) between two adjacent genes in a cluster, if no correlation is detected between the two genes (Default: 2500)\n",
        "max_bgc_gap = 15000 #@param {type:\"integer\"}\n",
        "#@markdown - Maximum allowed gap (in bp) between two adjacent genes, if they only have a weak correlation (Default: 15000)\n",
        "min_prot_len = 200 #@param {type:\"integer\"}\n",
        "#@markdown - Minimum protein length considered for biosynthetic protein detection (Default: 200)\n",
        "num_of_genes_checked = 20 #@param {type:\"integer\"}\n",
        "#@markdown - Number of genes around the core/target genes to be checked (Default: 20)\n",
        "min_identity = 50 #@param {type:\"integer\"}\n",
        "#@markdown - Minimum identity (in %) required to be considered as a duplicated protein (Default: 50)\n",
        "additional_genes = 0 #@param {type:\"integer\"}\n",
        "#@markdown - Number of additional genes to be included outside a predicted BGC (Default: 0)\n",
        "noCore = False #@param {type:\"boolean\"}\n",
        "#@markdown - Tick here if you wish to extract BGCs without a core protein as well. This option must be used with either `target` or `pfam` mode.\n",
        "noUstY = False #@param {type:\"boolean\"}\n",
        "#@markdown - Tick here if you wish to extract RiPP BGCs without a UstY homologue(s).\n",
        "noKexB = False #@param {type:\"boolean\"}\n",
        "#@markdown - Tick here if you wish to extract RiPP BGCs without a precursor peptide with KexB cleavage sites.\n",
        "NeedCSV = False #@param {type:\"boolean\"}\n",
        "#@markdown - Tick here if you wish to to include analysis CSV files in the output."
      ],
      "metadata": {
        "cellView": "form",
        "id": "j_ikcPuph4_C"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Install dependencies\n",
        "import os\n",
        "import shutil\n",
        "!sudo apt-get install hmmer diamond-aligner\n",
        "!pip install biopython\n",
        "!pip install natsort\n",
        "!pip install tensorflow\n",
        "if os.path.isdir(\"/content/funbgcex\") == False:\n",
        "  !git clone https://github.com/ydmatsd/funbgcex-colab2.git\n",
        "  shutil.move(\"/content/funbgcex-colab2/funbgcex-colab/funbgcex\",\"/content\")"
      ],
      "metadata": {
        "cellView": "form",
        "id": "um2ou0yLLBJ_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Extract BGCs\n",
        "import os\n",
        "import sys\n",
        "from funbgcex.BGCeXtractor import BGCeXtractorMain\n",
        "\n",
        "funbgcex_path = \"/content/funbgcex\"\n",
        "\n",
        "#Accept input files\n",
        "if os.path.isdir(gbk_dir) == False:\n",
        "    sys.exit(\"The input directory does not exist.\")\n",
        "if (mode == \"all\" or mode == \"ripps\" or mode == \"sre\" or mode == \"human\") and noCore:\n",
        "    sys.exit(\"'no_core' can only be used with the target or pfam mode.\")\n",
        "if additional_genes < 0:\n",
        "    sys.exit(\"Please provide a non-negative int value for additional_genes.\")\n",
        "if mode == \"all\" or mode == \"ripps\" or mode == \"sre\" or mode == \"human\":\n",
        "    query = \"none\"\n",
        "elif mode == \"target\":\n",
        "    query = fasta_path\n",
        "    if query == \"none\":\n",
        "        sys.exit(\"Please provide a fasta file.\")\n",
        "    if os.path.isfile(query) == False:\n",
        "        sys.exit(\"The fasta file does not exist.\")\n",
        "elif mode == \"pfam\":\n",
        "    PfamList = []\n",
        "    allHMM = f\"{funbgcex_path}/data/hmm/fungalSM/fungalSM.hmm\"\n",
        "    with open(allHMM,\"r\") as hmm:\n",
        "        for line in hmm:\n",
        "            if line.startswith(\"NAME\"):\n",
        "                name = line.split(\" \")[-1].replace(\"\\n\",\"\")\n",
        "                PfamList.append(name)\n",
        "    query = pfam\n",
        "    if query not in PfamList:\n",
        "        if query != \"none\":\n",
        "            print(f\"{query} does not exist in the database.\")\n",
        "            ToBeContinued = True\n",
        "            while ToBeContinued:\n",
        "                answer = input(\"Do you want to check the available protein family names? y/n: \")\n",
        "                if answer == \"Y\" or answer == \"y\":\n",
        "                    print(PfamList)\n",
        "                    ToBeContinued = False\n",
        "                    sys.exit()\n",
        "                elif answer == \"N\" or answer == \"n\":\n",
        "                    ToBeContinued = False\n",
        "                else:\n",
        "                    pass\n",
        "            sys.exit()\n",
        "        else:\n",
        "            sys.exit(\"Please provide the name of a protein family to be used for BGC extraction.\")\n",
        "else:\n",
        "    sys.exit(\"Please choose 'all', 'ripps', 'target', 'pfam', 'sre', or 'human' as an extraction mode.\")\n",
        "\n",
        "\"\"\"\n",
        "Start BGC extraction\n",
        "\"\"\"\n",
        "workers = 1\n",
        "BGCeXtractorMain(gbk_dir,results_dir,mode,query,gap_allowed,max_bgc_gap,min_prot_len,num_of_genes_checked,min_identity,additional_genes,noCore,noUstY,noKexB,workers,NeedCSV)\n"
      ],
      "metadata": {
        "id": "qbYzxjDlyMCk",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Download results\n",
        "import shutil\n",
        "from google.colab import files\n",
        "\n",
        "shutil.make_archive(\"/content/results\", 'zip', results_dir)\n",
        "shutil.rmtree(results_dir)\n",
        "shutil.rmtree(gbk_dir)\n",
        "files.download(\"/content/results.zip\")"
      ],
      "metadata": {
        "id": "YHHSUKW_2wNb",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "3TYvM043bVYa"
      }
    }
  ]
}