{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "4f2e2bfcccb34f349fd924570d33a51c": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ComboboxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ComboboxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ComboboxView",
            "continuous_update": true,
            "description": "Pfam:",
            "description_tooltip": null,
            "disabled": false,
            "ensure_option": true,
            "layout": "IPY_MODEL_901da8b871394471ab503815fd78289c",
            "options": [
              "2-Hacid_dh",
              "2-Hacid_dh_C",
              "2OG-FeII_Oxy",
              "2OG-FeII_Oxy_2",
              "2OG-FeII_Oxy_3",
              "2OG-Fe_Oxy_2",
              "3Beta_HSD",
              "3HCDH",
              "3HCDH_N",
              "4HBT",
              "4HBT_2",
              "4HBT_3",
              "AA-adenyl-dom",
              "AAA_16",
              "AAA_19",
              "AAA_21",
              "AAA_22",
              "AAA_23",
              "AAA_25",
              "AAA_29",
              "AAA_30",
              "AAA_33",
              "AAT",
              "AATase",
              "AA_kinase",
              "AA_permease",
              "AA_permease_2",
              "ABA3",
              "ABC1",
              "ABC2_membrane",
              "ABC2_membrane_3",
              "ABC2_membrane_7",
              "ABC_ATPase",
              "ABC_membrane",
              "ABC_tran",
              "ABC_trans_N",
              "ABM",
              "ACAS_N",
              "ACC_central",
              "ACPS",
              "ACP_PCP",
              "ACP_syn_III_C",
              "ACT",
              "ACT_7",
              "ADC",
              "ADH_N",
              "ADH_N_2",
              "ADH_zinc_N",
              "ADH_zinc_N_2",
              "ADP_ribosyl_GH",
              "ADSL_C",
              "AIG2_2",
              "APH",
              "ATEG_03636",
              "ATP-grasp",
              "ATP-grasp_4",
              "ATP-synt_ab",
              "ATP-synt_ab_N",
              "ATP11",
              "ATPgrasp_N",
              "ATPgrasp_Ter",
              "AXE1",
              "Aa_trans",
              "Abhydrolase_1",
              "Abhydrolase_2",
              "Abhydrolase_3",
              "Abhydrolase_6",
              "Acetyltransf_1",
              "Acetyltransf_3",
              "Acetyltransf_7",
              "Acetyltransf_8",
              "Aconitase",
              "Aconitase_C",
              "Acyl-CoA_dh_1",
              "Acyl-CoA_dh_2",
              "Acyl-CoA_dh_M",
              "Acyl-CoA_dh_N",
              "Acyl_transf_1",
              "Acyl_transf_3",
              "AflR",
              "Aldedh",
              "Aldo_ket_red",
              "Aldolase_II",
              "Aldose_epim",
              "Alpha-amylase",
              "Amidase",
              "Amidohydro_1",
              "Amidohydro_2",
              "Amidohydro_3",
              "Amino_oxidase",
              "Aminotran_1_2",
              "Aminotran_3",
              "Aminotran_4",
              "Aminotran_5",
              "Ank",
              "AnkA",
              "AnkG",
              "Ank_2",
              "Ank_3",
              "Ank_4",
              "Ank_5",
              "Anth_synt_I_N",
              "Anthrone_oxy",
              "ApbA",
              "ApbA_C",
              "Arginosuc_synth",
              "AsR5",
              "AsR6",
              "Asn_synthase",
              "Asp",
              "Asp_Glu_race",
              "Asp_aminotransf",
              "AstC",
              "AtaL",
              "AtnE",
              "BAAT_C",
              "BBE",
              "BD-FAE",
              "Bac_luciferase",
              "Beta-lactamase",
              "Beta_elim_lyase",
              "BfL6",
              "BfL8",
              "Biotin_carb_C",
              "Biotin_carb_N",
              "Biotin_lipoyl",
              "CBM-like",
              "CBS",
              "CMAS",
              "CMD",
              "COesterase",
              "CPSase_L_D2",
              "CPSase_L_D3",
              "CalD",
              "Caps_synth",
              "Carboxyl_trans",
              "Carn_acyltransf",
              "Catalase-rel",
              "Catalase",
              "CghF",
              "Chal_sti_synt_C",
              "Chal_sti_synt_N",
              "Chlorophyllase2",
              "Chorismate_bind",
              "Citrate_synt",
              "Clz3",
              "Clz7",
              "CoA_transf_3",
              "Cofilin_ADF",
              "Condensation",
              "CosA",
              "Cp1D",
              "CpaE",
              "CtvD",
              "Cu-oxidase",
              "Cu-oxidase_2",
              "Cu-oxidase_3",
              "Cu_amine_oxid",
              "Cupin_2",
              "Cys_Met_Meta_PP",
              "Cyt-b5",
              "Cytochrom_B561",
              "DAHP_synth_1",
              "DAO",
              "DDE_1",
              "DDE_3",
              "DHDPS",
              "DIOX_N",
              "DIT1_PvcA",
              "DJ-1_PfpI",
              "DLH",
              "DSBA",
              "DUF1295",
              "DUF1479",
              "DUF1729",
              "DUF1749",
              "DUF1752",
              "DUF1772",
              "DUF1776",
              "DUF1857",
              "DUF1929",
              "DUF1965",
              "DUF1993",
              "DUF2306",
              "DUF2334",
              "DUF2470",
              "DUF2784",
              "DUF3176",
              "DUF3237",
              "DUF3425",
              "DUF3471",
              "DUF3533",
              "DUF4149",
              "DUF4185",
              "DUF4188",
              "DUF4267",
              "DUF4385",
              "DUF4413",
              "DUF4451",
              "DUF5519",
              "DUF6534",
              "DUF6570",
              "DUF659",
              "DUF87",
              "Dabb",
              "Dala_Dala_lig_C",
              "Dimer_Tnp_hAT",
              "Dimerisation",
              "Dimerisation2",
              "Dynamin_N",
              "Dyp_perox",
              "ECH_1",
              "ECH_2",
              "EHN",
              "ERG4_ERG24",
              "EamA",
              "EcKL",
              "EcdJ",
              "EfHP",
              "Egh16-like",
              "ElcH",
              "Epimerase",
              "EqxR",
              "Esterase",
              "EthD",
              "F420_oxidored",
              "F5_F8_type_C",
              "FAA_hydrolase",
              "FAD_binding_1",
              "FAD_binding_2",
              "FAD_binding_3",
              "FAD_binding_4",
              "FAD_binding_6",
              "FAD_oxidored",
              "FAE1_CUT1_RppA",
              "FAS_I_H",
              "FAS_N",
              "FAS_meander",
              "FA_hydroxylase",
              "FMN_dh",
              "FMN_red",
              "FMO-like",
              "FPSE_09189",
              "FR47",
              "FSH1",
              "Fas_alpha_ACP",
              "Fasciclin",
              "Fe-ADH",
              "Fe-ADH_2",
              "Ferric_reduct",
              "FhuF",
              "Flavodoxin_1",
              "Flavodoxin_2",
              "Fn3-like",
              "FncF",
              "FolB",
              "FrsA-like",
              "FrzJ",
              "Fsa2",
              "FscB",
              "FtsJ",
              "FtsK_SpoIIIE",
              "Fungal_trans",
              "Fungal_trans_2",
              "GATase",
              "GATase_6",
              "GATase_7",
              "GBBH-like_N",
              "GDP_Man_Dehyd",
              "GFO_IDH_MocA",
              "GFO_IDH_MocA_C",
              "GGPS",
              "GH3",
              "GH43_C2",
              "GIDA",
              "GMC_oxred_C",
              "GMC_oxred_N",
              "GO-like_E_set",
              "GST_C",
              "GST_C_2",
              "GST_C_3",
              "GST_C_5",
              "GST_N",
              "GST_N_2",
              "GST_N_3",
              "GST_N_4",
              "GTP_cyclohydroI",
              "G_glu_transpept",
              "GliK",
              "GliN",
              "Gln-synt_C",
              "Glu_dehyd_C",
              "Gly_transf_sug",
              "Glyco_hydro_11",
              "Glyco_hydro_3",
              "Glyco_hydro_31",
              "Glyco_hydro_3_C",
              "Glyco_hydro_43",
              "Glyco_hydro_cc",
              "Glyco_tran_28_C",
              "Glyco_transf_28",
              "Glyoxal_oxid_N",
              "Glyoxalase",
              "Glyoxalase_3",
              "Glyoxalase_4",
              "Glyoxalase_6",
              "Gp_dh_C",
              "Gp_dh_N",
              "Gpr1_Fun34_YaaH",
              "GsfR1",
              "HAD_2",
              "HD",
              "HI0933_like",
              "HLH",
              "HMG-CoA_red",
              "HMGL-like",
              "HMG_CoA_synt_C",
              "HMG_CoA_synt_N",
              "HPIH",
              "HSP20",
              "HTH_51",
              "HTH_Tnp_Tc5",
              "HTH_psq",
              "Haem_degrading",
              "HbpS-like",
              "Helitron_like_N",
              "Helo_like_N",
              "Hemerythrin",
              "Hemocyanin_C",
              "Hemocyanin_M",
              "Hexapep",
              "HimD",
              "Histone",
              "Histone_H2A_C",
              "Homeobox_KN",
              "Homeodomain",
              "Homoserine_dh",
              "HpcH_HpaI",
              "Hydrolase",
              "Hydrolase_4",
              "Hydrolase_like",
              "ICMT",
              "IDO",
              "ILVD_EDD",
              "IMPDH",
              "IPPT",
              "IPT",
              "Iso_dh",
              "Isochorismatase",
              "IucA_IucC",
              "K_oxygenase",
              "Kdo",
              "Kelch_1",
              "Kelch_2",
              "Kelch_3",
              "Kelch_4",
              "Kelch_6",
              "LCM",
              "LIP",
              "Lactamase_B",
              "Lactamase_B_2",
              "Lactamase_B_3",
              "LcsT",
              "LigB",
              "Lipocalin_5",
              "Lipocalin_9",
              "LolU",
              "LtmS",
              "Lumazine_bd_2",
              "Lyase_1",
              "Lyase_aromatic",
              "Lycopene_cycl",
              "Lys_Orn_oxgnase",
              "Lysine_decarbox",
              "M20_dimer",
              "MARVEL",
              "MBOAT_2",
              "MFS_1",
              "MFS_1_like",
              "MFS_2",
              "MFS_4",
              "MFS_5",
              "MMR_HSR1",
              "MPAB_Lcp_cat",
              "MTS",
              "MULE",
              "MafB19-deam",
              "MaoC_dehydrat_N",
              "MaoC_dehydratas",
              "MdpA",
              "Melibiase_2",
              "Melibiase_C",
              "MetW",
              "Methyltransf_10",
              "Methyltransf_11",
              "Methyltransf_12",
              "Methyltransf_16",
              "Methyltransf_2",
              "Methyltransf_21",
              "Methyltransf_23",
              "Methyltransf_24",
              "Methyltransf_25",
              "Methyltransf_28",
              "Methyltransf_3",
              "Methyltransf_31",
              "Methyltransf_32",
              "Methyltransf_33",
              "Methyltransf_4",
              "Methyltransf_7",
              "Methyltransf_9",
              "Mito_carr",
              "MmgE_PrpD",
              "MmgE_PrpD_C",
              "MscL",
              "Myb_DNA-bind_6",
              "Myb_DNA-bind_7",
              "Myb_DNA-binding",
              "NACHT",
              "NAD_binding_1",
              "NAD_binding_10",
              "NAD_binding_11",
              "NAD_binding_2",
              "NAD_binding_3",
              "NAD_binding_4",
              "NAD_binding_6",
              "NAD_binding_8",
              "NAD_binding_9",
              "NMO",
              "NRPS_C",
              "NRPS_CT",
              "NRPS_E",
              "NRPS_term_dom",
              "Nitroreductase",
              "NmrA",
              "NnrU",
              "NodS",
              "NtCtMGAM_N",
              "NvfI",
              "O-FucT",
              "OATP",
              "OPT",
              "OphC",
              "Orn_Arg_deC_N",
              "Orn_DAP_Arg_deC",
              "OtaR1",
              "Oxidored_FMN",
              "Oxysterol_BP",
              "P5CR_dimer",
              "PAP2_3",
              "PBP",
              "PCMT",
              "PDR_CDR",
              "PEMT",
              "PEP_mutase",
              "PHM7_cyt",
              "PIF1",
              "PIGO_PIGG",
              "PKS_AT",
              "PKS_KS",
              "PKS_NRPS_MT",
              "PK_Tyr_Ser-Thr",
              "PPR_2",
              "PSDC",
              "PS_Dcarbxylase",
              "PTR2",
              "P_proprotein",
              "PatF",
              "PatJ",
              "Patched",
              "PaxC",
              "PbcA",
              "Pec_lyase",
              "PenF",
              "Peptidase_M19",
              "Peptidase_M20",
              "Peptidase_M24",
              "Peptidase_M3",
              "Peptidase_S10",
              "Peptidase_S15",
              "Peptidase_S28",
              "Peptidase_S41",
              "Peptidase_S8",
              "Peptidase_S9",
              "Peptidase_S9_N",
              "Peroxidase_2",
              "PfB",
              "Phe_hydrox_dim",
              "PhomB",
              "Phosphodiest",
              "PhyH",
              "Pkinase",
              "Polyketide_cyc2",
              "Polysacc_deac_1",
              "Polysacc_synt_2",
              "Prefoldin",
              "Prenyltrans",
              "Pribosyltran",
              "Pro_CA",
              "Pro_isomerase",
              "Proteasome",
              "PrpF",
              "PydY",
              "Pyr4",
              "Pyr_redox",
              "Pyr_redox_2",
              "Pyr_redox_3",
              "Pyridoxal_deC",
              "Questin_oxidase",
              "RSN1_7TM",
              "RSN1_TM",
              "RTA1",
              "RmlD_sub_bind",
              "RsgA_GTPase",
              "Rua1_C",
              "SAT2",
              "SET",
              "SGL",
              "SMC_N",
              "SQHop_cyclase_C",
              "SQHop_cyclase_N",
              "SQS_PSY",
              "Sacchrp_dh_NADP",
              "SalTPS",
              "Say1_Mug180",
              "Scytalone_dh",
              "Sdh_cyt",
              "SdnG",
              "SdnJ",
              "Semialdhyde_dh",
              "Shikimate_DH",
              "SmbD",
              "SnoaL",
              "SnoaL_2",
              "SnoaL_4",
              "Spd14",
              "SrdD",
              "Steroid_dh",
              "Sterol-sensing",
              "Sterol_MT_C",
              "Str_synth",
              "Sugar_tr",
              "Sulfotransfer_4",
              "T2SSE",
              "TFIIE_alpha",
              "TIM",
              "TPMT",
              "TPP_enzyme_C",
              "TPP_enzyme_M",
              "TPP_enzyme_N",
              "TPR_2",
              "TP_methylase",
              "TRAM1",
              "TRAM_LAG1_CLN8",
              "TRI12",
              "TRI5",
              "TRI9",
              "TTL",
              "Tannase",
              "TauD",
              "Tcf25",
              "TdiC",
              "Terpene_syn_C_2",
              "Terpene_synth_C",
              "Thi4",
              "Thioesterase",
              "Thiolase_N",
              "Tmemb_14",
              "TqaK",
              "Transferase",
              "Transketolase_C",
              "Transketolase_N",
              "Transp_cyt_pur",
              "Tri14",
              "Tri3",
              "Tri6",
              "Trp_DMAT",
              "Trp_halogenase",
              "Trt14",
              "Trt5",
              "Trypsin",
              "Trypsin_2",
              "TwmC",
              "Tyrosinase",
              "UDPGT",
              "UNC-93",
              "UbiA_PT",
              "UbiA_TC",
              "Ubie_methyltran",
              "UstYa",
              "VicA",
              "VicYx",
              "VidS",
              "VniA",
              "VrtF",
              "WD40",
              "XenC",
              "XenD",
              "YCII",
              "Zn_clus",
              "adh_short",
              "adh_short_C2",
              "bZIP_1",
              "dCMP_cyt_deam_1",
              "fPKS_DH",
              "fPKS_ER",
              "fPKS_KR",
              "fPKS_PT",
              "fPKS_R",
              "fPKS_TH",
              "fn3_3",
              "p450",
              "peroxidase",
              "tRNA-synt_1c",
              "tRNA-synt_1c_C",
              "tRNA-synt_2",
              "tRNA-synt_2c",
              "tRNA-synt_2d",
              "tRNA_anti-codon",
              "zf-C2H2",
              "zf-C6H2",
              "zf-H2C2_2",
              "zf_CCCH_4"
            ],
            "placeholder": "Type to search Pfam...",
            "style": "IPY_MODEL_9bebeb1f8edc4bc0ae93a512796af9ed",
            "value": "Pyr4"
          }
        },
        "901da8b871394471ab503815fd78289c": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "9bebeb1f8edc4bc0ae93a512796af9ed": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "3f4e500ed9b2498cae10bed9007ed824": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ButtonModel",
          "model_module_version": "1.5.0",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ButtonModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ButtonView",
            "button_style": "",
            "description": "Confirm",
            "disabled": false,
            "icon": "",
            "layout": "IPY_MODEL_260f6d4c62364dd9af6f6ab289e34f37",
            "style": "IPY_MODEL_f4eca612f2084c958ed9a1da368f9706",
            "tooltip": ""
          }
        },
        "260f6d4c62364dd9af6f6ab289e34f37": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "f4eca612f2084c958ed9a1da368f9706": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ButtonStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ButtonStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "button_color": null,
            "font_weight": ""
          }
        }
      }
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/ydmatsd/funbgcex-colab2/blob/main/funbgcex-colab/FunBGCeX.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "<img src=\"https://github.com/ydmatsd/funbgcex-colab/blob/main/funbgcex-colab/logo.png?raw=true\" height=\"60\" align=\"top\">\n",
        "\n",
        "##FunBGCeX (Fungal Biosynthetic Gene Cluster eXtractor) â€“ Google Colab Version\n",
        "[**FunBGCeX**](https://github.com/ydmatsd/funbgcex) is a genome mining tool for fungal natural product discovery, which detects biosynthetic proteins based on our manually curated database of fungal natural product biosynthetic gene clusters (BGCs), [FunBGCs](http://staffweb1.cityu.edu.hk/ymatsuda/funbgcs/funbgcs.html). FunBGCeX can extract not only all possible BGCs from given fungal genomes but also those encoding a protein of user's interest. The extracted BGCs are provided in the GenBank format and can also be visualized using a web browser (A sample output can be found [here](http://staffweb1.cityu.edu.hk/ymatsuda/funbgcex/sample/results.html)).\n",
        "### Citation\n",
        "[Tang, J., Matsuda, Y. *Nat. Commun.* **15**, 4312 (2024)](https://www.nature.com/articles/s41467-024-48771-7)\n",
        "\n",
        "-----------"
      ],
      "metadata": {
        "id": "K2KgE-a8bg6x"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Please provide information for BGC extraction, and then hit `Runtime` -> `Run all`\n",
        "UseDrive = False #@param {type:\"boolean\"}\n",
        "#@markdown - Direct file upload to Google Colab may take a long time. Tick here if you use a directory in your Google Drive (the path to the directory will be `/content/drive/MyDrive/[your directory name]`)\n",
        "if UseDrive:\n",
        "  from google.colab import drive\n",
        "  drive.mount('/content/drive')\n",
        "\n",
        "gbk_dir = \"/content/drive/MyDrive/YourDirectoryName\" #@param {type:\"string\"}\n",
        "#@markdown - Upload a directory that contains GenBank file(s) from which BGCs will be extracted and provide the path to the directory. The GenBank files need to have CDS features with a translation qualifier. Do not include any non-GenBank files in the directory.\n",
        "results_dir = \"/content/result\"\n",
        "mode = \"all\" #@param [\"all\", \"target\", \"pfam\"]\n",
        "#@markdown - Choose a mode for BGC extraction.\n",
        "#@markdown  - `all`: Extract all possible BGCs.\n",
        "#@markdown  - `target`: Extract BGCs that encode a homologue of protein(s) of your interest (a FASTA file is required).\n",
        "#@markdown  - `pfam`: Extract BGCs that encode a protein with a protein family/domain (Pfam) of your interest.\n",
        "fasta = \"content/YourProteins.fasta\" #@param {type:\"string\"}\n",
        "#@markdown - Required when `target` mode is used. Provide the path to a fasta file containing protein(s) of your interest.\n",
        "pfam = \"Pyr4\" #@param ['2-Hacid_dh', '2-Hacid_dh_C', '2OG-FeII_Oxy', '2OG-FeII_Oxy_2', '2OG-FeII_Oxy_3', '2OG-Fe_Oxy_2', '3Beta_HSD', '3HCDH', '3HCDH_N', '4HBT', '4HBT_2', '4HBT_3', 'AA-adenyl-dom', 'AAA_16', 'AAA_19', 'AAA_21', 'AAA_22', 'AAA_23', 'AAA_25', 'AAA_29', 'AAA_30', 'AAA_33', 'AAT', 'AATase', 'AA_kinase', 'AA_permease', 'AA_permease_2', 'ABA3', 'ABC1', 'ABC2_membrane', 'ABC2_membrane_3', 'ABC2_membrane_7', 'ABC_ATPase', 'ABC_membrane', 'ABC_tran', 'ABC_trans_N', 'ABM', 'ACAS_N', 'ACC_central', 'ACPS', 'ACP_PCP', 'ACP_syn_III_C', 'ACT', 'ACT_7', 'ADC', 'ADH_N', 'ADH_N_2', 'ADH_zinc_N', 'ADH_zinc_N_2', 'ADP_ribosyl_GH', 'ADSL_C', 'AIG2_2', 'APH', 'ATEG_03636', 'ATP-grasp', 'ATP-grasp_4', 'ATP-synt_ab', 'ATP-synt_ab_N', 'ATP11', 'ATPgrasp_N', 'ATPgrasp_Ter', 'AXE1', 'Aa_trans', 'Abhydrolase_1', 'Abhydrolase_2', 'Abhydrolase_3', 'Abhydrolase_6', 'Acetyltransf_1', 'Acetyltransf_3', 'Acetyltransf_7', 'Acetyltransf_8', 'Aconitase', 'Aconitase_C', 'Acyl-CoA_dh_1', 'Acyl-CoA_dh_2', 'Acyl-CoA_dh_M', 'Acyl-CoA_dh_N', 'Acyl_transf_1', 'Acyl_transf_3', 'AflR', 'Aldedh', 'Aldo_ket_red', 'Aldolase_II', 'Aldose_epim', 'Alpha-amylase', 'Amidase', 'Amidohydro_1', 'Amidohydro_2', 'Amidohydro_3', 'Amino_oxidase', 'Aminotran_1_2', 'Aminotran_3', 'Aminotran_4', 'Aminotran_5', 'Ank', 'AnkA', 'AnkG', 'Ank_2', 'Ank_3', 'Ank_4', 'Ank_5', 'Anth_synt_I_N', 'Anthrone_oxy', 'ApbA', 'ApbA_C', 'Arginosuc_synth', 'AsR5', 'AsR6', 'Asn_synthase', 'Asp', 'Asp_Glu_race', 'Asp_aminotransf', 'AstC', 'AtaL', 'AtnE', 'BAAT_C', 'BBE', 'BD-FAE', 'Bac_luciferase', 'Beta-lactamase', 'Beta_elim_lyase', 'BfL6', 'BfL8', 'Biotin_carb_C', 'Biotin_carb_N', 'Biotin_lipoyl', 'CBM-like', 'CBS', 'CMAS', 'CMD', 'COesterase', 'CPSase_L_D2', 'CPSase_L_D3', 'CalD', 'Caps_synth', 'Carboxyl_trans', 'Carn_acyltransf', 'Catalase-rel', 'Catalase', 'CghF', 'Chal_sti_synt_C', 'Chal_sti_synt_N', 'Chlorophyllase2', 'Chorismate_bind', 'Citrate_synt', 'Clz3', 'Clz7', 'CoA_transf_3', 'Cofilin_ADF', 'Condensation', 'CosA', 'Cp1D', 'CpaE', 'CtvD', 'Cu-oxidase', 'Cu-oxidase_2', 'Cu-oxidase_3', 'Cu_amine_oxid', 'Cupin_2', 'Cys_Met_Meta_PP', 'Cyt-b5', 'Cytochrom_B561', 'DAHP_synth_1', 'DAO', 'DDE_1', 'DDE_3', 'DHDPS', 'DIOX_N', 'DIT1_PvcA', 'DJ-1_PfpI', 'DLH', 'DSBA', 'DUF1295', 'DUF1479', 'DUF1729', 'DUF1749', 'DUF1752', 'DUF1772', 'DUF1776', 'DUF1857', 'DUF1929', 'DUF1965', 'DUF1993', 'DUF2306', 'DUF2334', 'DUF2470', 'DUF2784', 'DUF3176', 'DUF3237', 'DUF3425', 'DUF3471', 'DUF3533', 'DUF4149', 'DUF4185', 'DUF4188', 'DUF4267', 'DUF4385', 'DUF4413', 'DUF4451', 'DUF5519', 'DUF6534', 'DUF6570', 'DUF659', 'DUF87', 'Dabb', 'Dala_Dala_lig_C', 'Dimer_Tnp_hAT', 'Dimerisation', 'Dimerisation2', 'Dynamin_N', 'Dyp_perox', 'ECH_1', 'ECH_2', 'EHN', 'ERG4_ERG24', 'EamA', 'EcKL', 'EcdJ', 'EfHP', 'Egh16-like', 'ElcH', 'Epimerase', 'EqxR', 'Esterase', 'EthD', 'F420_oxidored', 'F5_F8_type_C', 'FAA_hydrolase', 'FAD_binding_1', 'FAD_binding_2', 'FAD_binding_3', 'FAD_binding_4', 'FAD_binding_6', 'FAD_oxidored', 'FAE1_CUT1_RppA', 'FAS_I_H', 'FAS_N', 'FAS_meander', 'FA_hydroxylase', 'FMN_dh', 'FMN_red', 'FMO-like', 'FPSE_09189', 'FR47', 'FSH1', 'Fas_alpha_ACP', 'Fasciclin', 'Fe-ADH', 'Fe-ADH_2', 'Ferric_reduct', 'FhuF', 'Flavodoxin_1', 'Flavodoxin_2', 'Fn3-like', 'FncF', 'FolB', 'FrsA-like', 'FrzJ', 'Fsa2', 'FscB', 'FtsJ', 'FtsK_SpoIIIE', 'Fungal_trans', 'Fungal_trans_2', 'GATase', 'GATase_6', 'GATase_7', 'GBBH-like_N', 'GDP_Man_Dehyd', 'GFO_IDH_MocA', 'GFO_IDH_MocA_C', 'GGPS', 'GH3', 'GH43_C2', 'GIDA', 'GMC_oxred_C', 'GMC_oxred_N', 'GO-like_E_set', 'GST_C', 'GST_C_2', 'GST_C_3', 'GST_C_5', 'GST_N', 'GST_N_2', 'GST_N_3', 'GST_N_4', 'GTP_cyclohydroI', 'G_glu_transpept', 'GliK', 'GliN', 'Gln-synt_C', 'Glu_dehyd_C', 'Gly_transf_sug', 'Glyco_hydro_11', 'Glyco_hydro_3', 'Glyco_hydro_31', 'Glyco_hydro_3_C', 'Glyco_hydro_43', 'Glyco_hydro_cc', 'Glyco_tran_28_C', 'Glyco_transf_28', 'Glyoxal_oxid_N', 'Glyoxalase', 'Glyoxalase_3', 'Glyoxalase_4', 'Glyoxalase_6', 'Gp_dh_C', 'Gp_dh_N', 'Gpr1_Fun34_YaaH', 'GsfR1', 'HAD_2', 'HD', 'HI0933_like', 'HLH', 'HMG-CoA_red', 'HMGL-like', 'HMG_CoA_synt_C', 'HMG_CoA_synt_N', 'HPIH', 'HSP20', 'HTH_51', 'HTH_Tnp_Tc5', 'HTH_psq', 'Haem_degrading', 'HbpS-like', 'Helitron_like_N', 'Helo_like_N', 'Hemerythrin', 'Hemocyanin_C', 'Hemocyanin_M', 'Hexapep', 'HimD', 'Histone', 'Histone_H2A_C', 'Homeobox_KN', 'Homeodomain', 'Homoserine_dh', 'HpcH_HpaI', 'Hydrolase', 'Hydrolase_4', 'Hydrolase_like', 'ICMT', 'IDO', 'ILVD_EDD', 'IMPDH', 'IPPT', 'IPT', 'Iso_dh', 'Isochorismatase', 'IucA_IucC', 'K_oxygenase', 'Kdo', 'Kelch_1', 'Kelch_2', 'Kelch_3', 'Kelch_4', 'Kelch_6', 'LCM', 'LIP', 'Lactamase_B', 'Lactamase_B_2', 'Lactamase_B_3', 'LcsT', 'LigB', 'Lipocalin_5', 'Lipocalin_9', 'LolU', 'LtmS', 'Lumazine_bd_2', 'Lyase_1', 'Lyase_aromatic', 'Lycopene_cycl', 'Lys_Orn_oxgnase', 'Lysine_decarbox', 'M20_dimer', 'MARVEL', 'MBOAT_2', 'MFS_1', 'MFS_1_like', 'MFS_2', 'MFS_4', 'MFS_5', 'MMR_HSR1', 'MPAB_Lcp_cat', 'MTS', 'MULE', 'MafB19-deam', 'MaoC_dehydrat_N', 'MaoC_dehydratas', 'MdpA', 'Melibiase_2', 'Melibiase_C', 'MetW', 'Methyltransf_10', 'Methyltransf_11', 'Methyltransf_12', 'Methyltransf_16', 'Methyltransf_2', 'Methyltransf_21', 'Methyltransf_23', 'Methyltransf_24', 'Methyltransf_25', 'Methyltransf_28', 'Methyltransf_3', 'Methyltransf_31', 'Methyltransf_32', 'Methyltransf_33', 'Methyltransf_4', 'Methyltransf_7', 'Methyltransf_9', 'Mito_carr', 'MmgE_PrpD', 'MmgE_PrpD_C', 'MscL', 'Myb_DNA-bind_6', 'Myb_DNA-bind_7', 'Myb_DNA-binding', 'NACHT', 'NAD_binding_1', 'NAD_binding_10', 'NAD_binding_11', 'NAD_binding_2', 'NAD_binding_3', 'NAD_binding_4', 'NAD_binding_6', 'NAD_binding_8', 'NAD_binding_9', 'NMO', 'NRPS_C', 'NRPS_CT', 'NRPS_E', 'NRPS_term_dom', 'Nitroreductase', 'NmrA', 'NnrU', 'NodS', 'NtCtMGAM_N', 'NvfI', 'O-FucT', 'OATP', 'OPT', 'OphC', 'Orn_Arg_deC_N', 'Orn_DAP_Arg_deC', 'OtaR1', 'Oxidored_FMN', 'Oxysterol_BP', 'P5CR_dimer', 'PAP2_3', 'PBP', 'PCMT', 'PDR_CDR', 'PEMT', 'PEP_mutase', 'PHM7_cyt', 'PIF1', 'PIGO_PIGG', 'PKS_AT', 'PKS_KS', 'PKS_NRPS_MT', 'PK_Tyr_Ser-Thr', 'PPR_2', 'PSDC', 'PS_Dcarbxylase', 'PTR2', 'P_proprotein', 'PatF', 'PatJ', 'Patched', 'PaxC', 'PbcA', 'Pec_lyase', 'PenF', 'Peptidase_M19', 'Peptidase_M20', 'Peptidase_M24', 'Peptidase_M3', 'Peptidase_S10', 'Peptidase_S15', 'Peptidase_S28', 'Peptidase_S41', 'Peptidase_S8', 'Peptidase_S9', 'Peptidase_S9_N', 'Peroxidase_2', 'PfB', 'Phe_hydrox_dim', 'PhomB', 'Phosphodiest', 'PhyH', 'Pkinase', 'Polyketide_cyc2', 'Polysacc_deac_1', 'Polysacc_synt_2', 'Prefoldin', 'Prenyltrans', 'Pribosyltran', 'Pro_CA', 'Pro_isomerase', 'Proteasome', 'PrpF', 'PydY', 'Pyr4', 'Pyr_redox', 'Pyr_redox_2', 'Pyr_redox_3', 'Pyridoxal_deC', 'Questin_oxidase', 'RSN1_7TM', 'RSN1_TM', 'RTA1', 'RmlD_sub_bind', 'RsgA_GTPase', 'Rua1_C', 'SAT2', 'SET', 'SGL', 'SMC_N', 'SQHop_cyclase_C', 'SQHop_cyclase_N', 'SQS_PSY', 'Sacchrp_dh_NADP', 'SalTPS', 'Say1_Mug180', 'Scytalone_dh', 'Sdh_cyt', 'SdnG', 'SdnJ', 'Semialdhyde_dh', 'Shikimate_DH', 'SmbD', 'SnoaL', 'SnoaL_2', 'SnoaL_4', 'Spd14', 'SrdD', 'Steroid_dh', 'Sterol-sensing', 'Sterol_MT_C', 'Str_synth', 'Sugar_tr', 'Sulfotransfer_4', 'T2SSE', 'TFIIE_alpha', 'TIM', 'TPMT', 'TPP_enzyme_C', 'TPP_enzyme_M', 'TPP_enzyme_N', 'TPR_2', 'TP_methylase', 'TRAM1', 'TRAM_LAG1_CLN8', 'TRI12', 'TRI5', 'TRI9', 'TTL', 'Tannase', 'TauD', 'Tcf25', 'TdiC', 'Terpene_syn_C_2', 'Terpene_synth_C', 'Thi4', 'Thioesterase', 'Thiolase_N', 'Tmemb_14', 'TqaK', 'Transferase', 'Transketolase_C', 'Transketolase_N', 'Transp_cyt_pur', 'Tri14', 'Tri3', 'Tri6', 'Trp_DMAT', 'Trp_halogenase', 'Trt14', 'Trt5', 'Trypsin', 'Trypsin_2', 'TwmC', 'Tyrosinase', 'UDPGT', 'UNC-93', 'UbiA_PT', 'UbiA_TC', 'Ubie_methyltran', 'UstYa', 'VicA', 'VicYx', 'VidS', 'VniA', 'VrtF', 'WD40', 'XenC', 'XenD', 'YCII', 'Zn_clus', 'adh_short', 'adh_short_C2', 'bZIP_1', 'dCMP_cyt_deam_1', 'fPKS_DH', 'fPKS_ER', 'fPKS_KR', 'fPKS_PT', 'fPKS_R', 'fPKS_TH', 'fn3_3', 'p450', 'peroxidase', 'tRNA-synt_1c', 'tRNA-synt_1c_C', 'tRNA-synt_2', 'tRNA-synt_2c', 'tRNA-synt_2d', 'tRNA_anti-codon', 'zf-C2H2', 'zf-C6H2', 'zf-H2C2_2', 'zf_CCCH_4']\n",
        "#@markdown - Required when `pfam` mode is used. Choose a protein domain of your interest.\n",
        "\n",
        "#@markdown **Optional settings**\n",
        "gap_allowed = 2500 #@param {type:\"integer\"}\n",
        "#@markdown - Maximum allowed gap (in bp) between two adjacent genes in a cluster, if no correlation is detected between the two genes (Default: 2500)\n",
        "max_bgc_gap = 15000 #@param {type:\"integer\"}\n",
        "#@markdown - Maximum allowed gap (in bp) between two adjacent genes, if they only have a weak correlation (Default: 15000)\n",
        "min_prot_len = 200 #@param {type:\"integer\"}\n",
        "#@markdown - Minimum protein length considered for biosynthetic protein detection (Default: 200)\n",
        "num_of_genes_checked = 20 #@param {type:\"integer\"}\n",
        "#@markdown - Number of genes around the core/target genes to be checked (Default: 20)\n",
        "min_identity = 50 #@param {type:\"integer\"}\n",
        "#@markdown - Minimum identity (in %) required to be considered as a duplicated protein (Default: 50)"
      ],
      "metadata": {
        "id": "T0DJuPWow2Ed"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Please provide information for BGC extraction, and then hit `Runtime` -> `Run all`\n",
        "\n",
        "import os\n",
        "import shutil\n",
        "from google.colab import files\n",
        "\n",
        "#Make a directory for direct upload\n",
        "gbk_dir = \"/content/gbk_dir\"\n",
        "os.makedirs(gbk_dir, exist_ok=True)\n",
        "\n",
        "UseDrive = False #@param {type:\"boolean\"}\n",
        "#@markdown - Direct file upload to Google Colab may take a long time. Tick here if you use a directory in your Google Drive (the path to the directory will be `/content/drive/MyDrive/[your directory name]`)\n",
        "if UseDrive:\n",
        "  from google.colab import drive\n",
        "  drive.mount('/content/drive')\n",
        "  gbk_dir = \"/content/drive/MyDrive/YourDirectoryName\" #@param {type:\"string\"}\n",
        "\n",
        "\n",
        "#@markdown - Upload a directory that contains GenBank file(s) from which BGCs will be extracted and provide the path to the directory. The GenBank files need to have CDS features with a translation qualifier. Do not include any non-GenBank files in the directory.\n",
        "results_dir = \"/content/result\"\n",
        "mode = \"all\" #@param [\"all\", \"target\", \"pfam\"]\n",
        "#@markdown - Choose a mode for BGC extraction.\n",
        "#@markdown  - `all`: Extract all possible BGCs.\n",
        "#@markdown  - `target`: Extract BGCs that encode a homologue of protein(s) of your interest (a FASTA file is required).\n",
        "#@markdown  - `pfam`: Extract BGCs that encode a protein with a protein family/domain (Pfam) of your interest.\n",
        "fasta = \"content/YourProteins.fasta\" #@param {type:\"string\"}\n",
        "#@markdown - Required when `target` mode is used. Provide the path to a fasta file containing protein(s) of your interest.\n",
        "pfam = \"Pyr4\" #@param ['2-Hacid_dh', '2-Hacid_dh_C', '2OG-FeII_Oxy', '2OG-FeII_Oxy_2', '2OG-FeII_Oxy_3', '2OG-Fe_Oxy_2', '3Beta_HSD', '3HCDH', '3HCDH_N', '4HBT', '4HBT_2', '4HBT_3', 'AA-adenyl-dom', 'AAA_16', 'AAA_19', 'AAA_21', 'AAA_22', 'AAA_23', 'AAA_25', 'AAA_29', 'AAA_30', 'AAA_33', 'AAT', 'AATase', 'AA_kinase', 'AA_permease', 'AA_permease_2', 'ABA3', 'ABC1', 'ABC2_membrane', 'ABC2_membrane_3', 'ABC2_membrane_7', 'ABC_ATPase', 'ABC_membrane', 'ABC_tran', 'ABC_trans_N', 'ABM', 'ACAS_N', 'ACC_central', 'ACPS', 'ACP_PCP', 'ACP_syn_III_C', 'ACT', 'ACT_7', 'ADC', 'ADH_N', 'ADH_N_2', 'ADH_zinc_N', 'ADH_zinc_N_2', 'ADP_ribosyl_GH', 'ADSL_C', 'AIG2_2', 'APH', 'ATEG_03636', 'ATP-grasp', 'ATP-grasp_4', 'ATP-synt_ab', 'ATP-synt_ab_N', 'ATP11', 'ATPgrasp_N', 'ATPgrasp_Ter', 'AXE1', 'Aa_trans', 'Abhydrolase_1', 'Abhydrolase_2', 'Abhydrolase_3', 'Abhydrolase_6', 'Acetyltransf_1', 'Acetyltransf_3', 'Acetyltransf_7', 'Acetyltransf_8', 'Aconitase', 'Aconitase_C', 'Acyl-CoA_dh_1', 'Acyl-CoA_dh_2', 'Acyl-CoA_dh_M', 'Acyl-CoA_dh_N', 'Acyl_transf_1', 'Acyl_transf_3', 'AflR', 'Aldedh', 'Aldo_ket_red', 'Aldolase_II', 'Aldose_epim', 'Alpha-amylase', 'Amidase', 'Amidohydro_1', 'Amidohydro_2', 'Amidohydro_3', 'Amino_oxidase', 'Aminotran_1_2', 'Aminotran_3', 'Aminotran_4', 'Aminotran_5', 'Ank', 'AnkA', 'AnkG', 'Ank_2', 'Ank_3', 'Ank_4', 'Ank_5', 'Anth_synt_I_N', 'Anthrone_oxy', 'ApbA', 'ApbA_C', 'Arginosuc_synth', 'AsR5', 'AsR6', 'Asn_synthase', 'Asp', 'Asp_Glu_race', 'Asp_aminotransf', 'AstC', 'AtaL', 'AtnE', 'BAAT_C', 'BBE', 'BD-FAE', 'Bac_luciferase', 'Beta-lactamase', 'Beta_elim_lyase', 'BfL6', 'BfL8', 'Biotin_carb_C', 'Biotin_carb_N', 'Biotin_lipoyl', 'CBM-like', 'CBS', 'CMAS', 'CMD', 'COesterase', 'CPSase_L_D2', 'CPSase_L_D3', 'CalD', 'Caps_synth', 'Carboxyl_trans', 'Carn_acyltransf', 'Catalase-rel', 'Catalase', 'CghF', 'Chal_sti_synt_C', 'Chal_sti_synt_N', 'Chlorophyllase2', 'Chorismate_bind', 'Citrate_synt', 'Clz3', 'Clz7', 'CoA_transf_3', 'Cofilin_ADF', 'Condensation', 'CosA', 'Cp1D', 'CpaE', 'CtvD', 'Cu-oxidase', 'Cu-oxidase_2', 'Cu-oxidase_3', 'Cu_amine_oxid', 'Cupin_2', 'Cys_Met_Meta_PP', 'Cyt-b5', 'Cytochrom_B561', 'DAHP_synth_1', 'DAO', 'DDE_1', 'DDE_3', 'DHDPS', 'DIOX_N', 'DIT1_PvcA', 'DJ-1_PfpI', 'DLH', 'DSBA', 'DUF1295', 'DUF1479', 'DUF1729', 'DUF1749', 'DUF1752', 'DUF1772', 'DUF1776', 'DUF1857', 'DUF1929', 'DUF1965', 'DUF1993', 'DUF2306', 'DUF2334', 'DUF2470', 'DUF2784', 'DUF3176', 'DUF3237', 'DUF3425', 'DUF3471', 'DUF3533', 'DUF4149', 'DUF4185', 'DUF4188', 'DUF4267', 'DUF4385', 'DUF4413', 'DUF4451', 'DUF5519', 'DUF6534', 'DUF6570', 'DUF659', 'DUF87', 'Dabb', 'Dala_Dala_lig_C', 'Dimer_Tnp_hAT', 'Dimerisation', 'Dimerisation2', 'Dynamin_N', 'Dyp_perox', 'ECH_1', 'ECH_2', 'EHN', 'ERG4_ERG24', 'EamA', 'EcKL', 'EcdJ', 'EfHP', 'Egh16-like', 'ElcH', 'Epimerase', 'EqxR', 'Esterase', 'EthD', 'F420_oxidored', 'F5_F8_type_C', 'FAA_hydrolase', 'FAD_binding_1', 'FAD_binding_2', 'FAD_binding_3', 'FAD_binding_4', 'FAD_binding_6', 'FAD_oxidored', 'FAE1_CUT1_RppA', 'FAS_I_H', 'FAS_N', 'FAS_meander', 'FA_hydroxylase', 'FMN_dh', 'FMN_red', 'FMO-like', 'FPSE_09189', 'FR47', 'FSH1', 'Fas_alpha_ACP', 'Fasciclin', 'Fe-ADH', 'Fe-ADH_2', 'Ferric_reduct', 'FhuF', 'Flavodoxin_1', 'Flavodoxin_2', 'Fn3-like', 'FncF', 'FolB', 'FrsA-like', 'FrzJ', 'Fsa2', 'FscB', 'FtsJ', 'FtsK_SpoIIIE', 'Fungal_trans', 'Fungal_trans_2', 'GATase', 'GATase_6', 'GATase_7', 'GBBH-like_N', 'GDP_Man_Dehyd', 'GFO_IDH_MocA', 'GFO_IDH_MocA_C', 'GGPS', 'GH3', 'GH43_C2', 'GIDA', 'GMC_oxred_C', 'GMC_oxred_N', 'GO-like_E_set', 'GST_C', 'GST_C_2', 'GST_C_3', 'GST_C_5', 'GST_N', 'GST_N_2', 'GST_N_3', 'GST_N_4', 'GTP_cyclohydroI', 'G_glu_transpept', 'GliK', 'GliN', 'Gln-synt_C', 'Glu_dehyd_C', 'Gly_transf_sug', 'Glyco_hydro_11', 'Glyco_hydro_3', 'Glyco_hydro_31', 'Glyco_hydro_3_C', 'Glyco_hydro_43', 'Glyco_hydro_cc', 'Glyco_tran_28_C', 'Glyco_transf_28', 'Glyoxal_oxid_N', 'Glyoxalase', 'Glyoxalase_3', 'Glyoxalase_4', 'Glyoxalase_6', 'Gp_dh_C', 'Gp_dh_N', 'Gpr1_Fun34_YaaH', 'GsfR1', 'HAD_2', 'HD', 'HI0933_like', 'HLH', 'HMG-CoA_red', 'HMGL-like', 'HMG_CoA_synt_C', 'HMG_CoA_synt_N', 'HPIH', 'HSP20', 'HTH_51', 'HTH_Tnp_Tc5', 'HTH_psq', 'Haem_degrading', 'HbpS-like', 'Helitron_like_N', 'Helo_like_N', 'Hemerythrin', 'Hemocyanin_C', 'Hemocyanin_M', 'Hexapep', 'HimD', 'Histone', 'Histone_H2A_C', 'Homeobox_KN', 'Homeodomain', 'Homoserine_dh', 'HpcH_HpaI', 'Hydrolase', 'Hydrolase_4', 'Hydrolase_like', 'ICMT', 'IDO', 'ILVD_EDD', 'IMPDH', 'IPPT', 'IPT', 'Iso_dh', 'Isochorismatase', 'IucA_IucC', 'K_oxygenase', 'Kdo', 'Kelch_1', 'Kelch_2', 'Kelch_3', 'Kelch_4', 'Kelch_6', 'LCM', 'LIP', 'Lactamase_B', 'Lactamase_B_2', 'Lactamase_B_3', 'LcsT', 'LigB', 'Lipocalin_5', 'Lipocalin_9', 'LolU', 'LtmS', 'Lumazine_bd_2', 'Lyase_1', 'Lyase_aromatic', 'Lycopene_cycl', 'Lys_Orn_oxgnase', 'Lysine_decarbox', 'M20_dimer', 'MARVEL', 'MBOAT_2', 'MFS_1', 'MFS_1_like', 'MFS_2', 'MFS_4', 'MFS_5', 'MMR_HSR1', 'MPAB_Lcp_cat', 'MTS', 'MULE', 'MafB19-deam', 'MaoC_dehydrat_N', 'MaoC_dehydratas', 'MdpA', 'Melibiase_2', 'Melibiase_C', 'MetW', 'Methyltransf_10', 'Methyltransf_11', 'Methyltransf_12', 'Methyltransf_16', 'Methyltransf_2', 'Methyltransf_21', 'Methyltransf_23', 'Methyltransf_24', 'Methyltransf_25', 'Methyltransf_28', 'Methyltransf_3', 'Methyltransf_31', 'Methyltransf_32', 'Methyltransf_33', 'Methyltransf_4', 'Methyltransf_7', 'Methyltransf_9', 'Mito_carr', 'MmgE_PrpD', 'MmgE_PrpD_C', 'MscL', 'Myb_DNA-bind_6', 'Myb_DNA-bind_7', 'Myb_DNA-binding', 'NACHT', 'NAD_binding_1', 'NAD_binding_10', 'NAD_binding_11', 'NAD_binding_2', 'NAD_binding_3', 'NAD_binding_4', 'NAD_binding_6', 'NAD_binding_8', 'NAD_binding_9', 'NMO', 'NRPS_C', 'NRPS_CT', 'NRPS_E', 'NRPS_term_dom', 'Nitroreductase', 'NmrA', 'NnrU', 'NodS', 'NtCtMGAM_N', 'NvfI', 'O-FucT', 'OATP', 'OPT', 'OphC', 'Orn_Arg_deC_N', 'Orn_DAP_Arg_deC', 'OtaR1', 'Oxidored_FMN', 'Oxysterol_BP', 'P5CR_dimer', 'PAP2_3', 'PBP', 'PCMT', 'PDR_CDR', 'PEMT', 'PEP_mutase', 'PHM7_cyt', 'PIF1', 'PIGO_PIGG', 'PKS_AT', 'PKS_KS', 'PKS_NRPS_MT', 'PK_Tyr_Ser-Thr', 'PPR_2', 'PSDC', 'PS_Dcarbxylase', 'PTR2', 'P_proprotein', 'PatF', 'PatJ', 'Patched', 'PaxC', 'PbcA', 'Pec_lyase', 'PenF', 'Peptidase_M19', 'Peptidase_M20', 'Peptidase_M24', 'Peptidase_M3', 'Peptidase_S10', 'Peptidase_S15', 'Peptidase_S28', 'Peptidase_S41', 'Peptidase_S8', 'Peptidase_S9', 'Peptidase_S9_N', 'Peroxidase_2', 'PfB', 'Phe_hydrox_dim', 'PhomB', 'Phosphodiest', 'PhyH', 'Pkinase', 'Polyketide_cyc2', 'Polysacc_deac_1', 'Polysacc_synt_2', 'Prefoldin', 'Prenyltrans', 'Pribosyltran', 'Pro_CA', 'Pro_isomerase', 'Proteasome', 'PrpF', 'PydY', 'Pyr4', 'Pyr_redox', 'Pyr_redox_2', 'Pyr_redox_3', 'Pyridoxal_deC', 'Questin_oxidase', 'RSN1_7TM', 'RSN1_TM', 'RTA1', 'RmlD_sub_bind', 'RsgA_GTPase', 'Rua1_C', 'SAT2', 'SET', 'SGL', 'SMC_N', 'SQHop_cyclase_C', 'SQHop_cyclase_N', 'SQS_PSY', 'Sacchrp_dh_NADP', 'SalTPS', 'Say1_Mug180', 'Scytalone_dh', 'Sdh_cyt', 'SdnG', 'SdnJ', 'Semialdhyde_dh', 'Shikimate_DH', 'SmbD', 'SnoaL', 'SnoaL_2', 'SnoaL_4', 'Spd14', 'SrdD', 'Steroid_dh', 'Sterol-sensing', 'Sterol_MT_C', 'Str_synth', 'Sugar_tr', 'Sulfotransfer_4', 'T2SSE', 'TFIIE_alpha', 'TIM', 'TPMT', 'TPP_enzyme_C', 'TPP_enzyme_M', 'TPP_enzyme_N', 'TPR_2', 'TP_methylase', 'TRAM1', 'TRAM_LAG1_CLN8', 'TRI12', 'TRI5', 'TRI9', 'TTL', 'Tannase', 'TauD', 'Tcf25', 'TdiC', 'Terpene_syn_C_2', 'Terpene_synth_C', 'Thi4', 'Thioesterase', 'Thiolase_N', 'Tmemb_14', 'TqaK', 'Transferase', 'Transketolase_C', 'Transketolase_N', 'Transp_cyt_pur', 'Tri14', 'Tri3', 'Tri6', 'Trp_DMAT', 'Trp_halogenase', 'Trt14', 'Trt5', 'Trypsin', 'Trypsin_2', 'TwmC', 'Tyrosinase', 'UDPGT', 'UNC-93', 'UbiA_PT', 'UbiA_TC', 'Ubie_methyltran', 'UstYa', 'VicA', 'VicYx', 'VidS', 'VniA', 'VrtF', 'WD40', 'XenC', 'XenD', 'YCII', 'Zn_clus', 'adh_short', 'adh_short_C2', 'bZIP_1', 'dCMP_cyt_deam_1', 'fPKS_DH', 'fPKS_ER', 'fPKS_KR', 'fPKS_PT', 'fPKS_R', 'fPKS_TH', 'fn3_3', 'p450', 'peroxidase', 'tRNA-synt_1c', 'tRNA-synt_1c_C', 'tRNA-synt_2', 'tRNA-synt_2c', 'tRNA-synt_2d', 'tRNA_anti-codon', 'zf-C2H2', 'zf-C6H2', 'zf-H2C2_2', 'zf_CCCH_4']\n",
        "#@markdown - Required when `pfam` mode is used. Choose a protein domain of your interest.\n",
        "\n",
        "#@markdown **Optional settings**\n",
        "gap_allowed = 2500 #@param {type:\"integer\"}\n",
        "#@markdown - Maximum allowed gap (in bp) between two adjacent genes in a cluster, if no correlation is detected between the two genes (Default: 2500)\n",
        "max_bgc_gap = 15000 #@param {type:\"integer\"}\n",
        "#@markdown - Maximum allowed gap (in bp) between two adjacent genes, if they only have a weak correlation (Default: 15000)\n",
        "min_prot_len = 200 #@param {type:\"integer\"}\n",
        "#@markdown - Minimum protein length considered for biosynthetic protein detection (Default: 200)\n",
        "num_of_genes_checked = 20 #@param {type:\"integer\"}\n",
        "#@markdown - Number of genes around the core/target genes to be checked (Default: 20)\n",
        "min_identity = 50 #@param {type:\"integer\"}\n",
        "#@markdown - Minimum identity (in %) required to be considered as a duplicated protein (Default: 50)"
      ],
      "metadata": {
        "id": "zSHbYBny7RVM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Please provide information for BGC extraction, and then hit `Runtime` -> `Run all`\n",
        "\n",
        "import os\n",
        "import shutil\n",
        "from google.colab import files\n",
        "\n",
        "\n",
        "InputMode = \"path\" #@param [\"upload\", \"path\"]\n",
        "#@markdown - Choose how to provide input data:\n",
        "#@markdown  - `upload`: Upload file(s) or a folder (zip).\n",
        "#@markdown  - `path`: Specify an existing folder path (Google Drive or Colab local). <b>Ensure that no non-GenBank files are placed in this folder.</b>\n",
        "\n",
        "#@markdown **Only for `upload` as InputMode**\n",
        "\n",
        "if InputMode == \"upload\":\n",
        "    UploadType = \"file\" #@param [\"file\", \"folder\"]\n",
        "    #@markdown - Choose upload type when using `upload`:\n",
        "    #@markdown  - `file`: Upload one or more GenBank files.\n",
        "    #@markdown  - `folder`: Upload a ZIP-compressed folder only containing GenBank files. <b>Ensure that no non-GenBank files are placed in this folder.</b>\n",
        "\n",
        "    gbk_dir = \"/content/gbk_dir\"\n",
        "    os.makedirs(gbk_dir, exist_ok=True)\n",
        "\n",
        "    if UploadType == \"file\":\n",
        "        print(\"ðŸ“‚ Please upload your GenBank file(s)\")\n",
        "        uploaded = files.upload()\n",
        "        for fname in uploaded.keys():\n",
        "            dest_path = os.path.join(gbk_dir, fname)\n",
        "            shutil.move(fname, dest_path)\n",
        "            print(f\"âœ… Uploaded file saved to {dest_path}\")\n",
        "    else:  # UploadType == \"folder\"\n",
        "        print(\"ðŸ“¦ Please upload a ZIP file containing your GenBank folder\")\n",
        "        uploaded = files.upload()\n",
        "        for fname in uploaded.keys():\n",
        "            zip_path = os.path.join(\"/content\", fname)\n",
        "            shutil.move(fname, zip_path)\n",
        "            os.system(f\"unzip -o {zip_path} -d {gbk_dir}\")\n",
        "            print(f\"âœ… Folder extracted to {gbk_dir}\")\n",
        "\n",
        "#@markdown **Only for `path` as InputMode**\n",
        "\n",
        "elif InputMode == \"path\":\n",
        "    from google.colab import drive\n",
        "    UseDrive = False #@param {type:\"boolean\"}\n",
        "    #@markdown - Tick here if you want to use Google Drive.\n",
        "\n",
        "    if UseDrive:\n",
        "        drive.mount('/content/drive')\n",
        "\n",
        "    gbk_dir = \"/content/gbk_dir\" #@param {type:\"string\"}\n",
        "    #@markdown - Or specify any existing folder path (e.g. `/content/my_data`).\n",
        "    print(f\"ðŸ“ Using existing folder: {gbk_dir}\")\n",
        "\n",
        "\n",
        "#@markdown - Upload a directory that contains GenBank file(s) from which BGCs will be extracted and provide the path to the directory. The GenBank files need to have CDS features with a translation qualifier. Do not include any non-GenBank files in the directory.\n",
        "results_dir = \"/content/result\"\n",
        "\n",
        "#@markdown **Choose a mode for BGC extraction**\n",
        "mode = \"pfam\" #@param [\"all\", \"target\", \"pfam\", \"ripps\", \"sre\", \"human\"]\n",
        "#@markdown - Choose a mode for BGC extraction.\n",
        "#@markdown  - `all`: Extract all possible BGCs.\n",
        "#@markdown  - `target`: Extract BGCs that encode a homologue of protein(s) of your interest (a FASTA file is required).\n",
        "#@markdown  - `pfam`: Extract BGCs that encode a protein with a protein family/domain (Pfam) of your interest.\n",
        "#@markdown  - `ripps`: Extract BGCs that encode a possible precursor peptide for RiPP biosynthesis.\n",
        "#@markdown  - `sre`: Extract BGCs with a possible self-resistance gene.\n",
        "#@markdown  - `human`: Extract BGCs that encode a homologue of human disease-related proteins.\n",
        "\n",
        "if mode == \"pfam\":\n",
        "  import ipywidgets as widgets\n",
        "  from IPython.display import display,clear_output\n",
        "\n",
        "  pfam_list = ['2-Hacid_dh', '2-Hacid_dh_C', '2OG-FeII_Oxy', '2OG-FeII_Oxy_2', '2OG-FeII_Oxy_3', '2OG-Fe_Oxy_2', '3Beta_HSD', '3HCDH', '3HCDH_N', '4HBT', '4HBT_2', '4HBT_3', 'AA-adenyl-dom', 'AAA_16', 'AAA_19', 'AAA_21', 'AAA_22', 'AAA_23', 'AAA_25', 'AAA_29', 'AAA_30', 'AAA_33', 'AAT', 'AATase', 'AA_kinase', 'AA_permease', 'AA_permease_2', 'ABA3', 'ABC1', 'ABC2_membrane', 'ABC2_membrane_3', 'ABC2_membrane_7', 'ABC_ATPase', 'ABC_membrane', 'ABC_tran', 'ABC_trans_N', 'ABM', 'ACAS_N', 'ACC_central', 'ACPS', 'ACP_PCP', 'ACP_syn_III_C', 'ACT', 'ACT_7', 'ADC', 'ADH_N', 'ADH_N_2', 'ADH_zinc_N', 'ADH_zinc_N_2', 'ADP_ribosyl_GH', 'ADSL_C', 'AIG2_2', 'APH', 'ATEG_03636', 'ATP-grasp', 'ATP-grasp_4', 'ATP-synt_ab', 'ATP-synt_ab_N', 'ATP11', 'ATPgrasp_N', 'ATPgrasp_Ter', 'AXE1', 'Aa_trans', 'Abhydrolase_1', 'Abhydrolase_2', 'Abhydrolase_3', 'Abhydrolase_6', 'Acetyltransf_1', 'Acetyltransf_3', 'Acetyltransf_7', 'Acetyltransf_8', 'Aconitase', 'Aconitase_C', 'Acyl-CoA_dh_1', 'Acyl-CoA_dh_2', 'Acyl-CoA_dh_M', 'Acyl-CoA_dh_N', 'Acyl_transf_1', 'Acyl_transf_3', 'AflR', 'Aldedh', 'Aldo_ket_red', 'Aldolase_II', 'Aldose_epim', 'Alpha-amylase', 'Amidase', 'Amidohydro_1', 'Amidohydro_2', 'Amidohydro_3', 'Amino_oxidase', 'Aminotran_1_2', 'Aminotran_3', 'Aminotran_4', 'Aminotran_5', 'Ank', 'AnkA', 'AnkG', 'Ank_2', 'Ank_3', 'Ank_4', 'Ank_5', 'Anth_synt_I_N', 'Anthrone_oxy', 'ApbA', 'ApbA_C', 'Arginosuc_synth', 'AsR5', 'AsR6', 'Asn_synthase', 'Asp', 'Asp_Glu_race', 'Asp_aminotransf', 'AstC', 'AtaL', 'AtnE', 'BAAT_C', 'BBE', 'BD-FAE', 'Bac_luciferase', 'Beta-lactamase', 'Beta_elim_lyase', 'BfL6', 'BfL8', 'Biotin_carb_C', 'Biotin_carb_N', 'Biotin_lipoyl', 'CBM-like', 'CBS', 'CMAS', 'CMD', 'COesterase', 'CPSase_L_D2', 'CPSase_L_D3', 'CalD', 'Caps_synth', 'Carboxyl_trans', 'Carn_acyltransf', 'Catalase-rel', 'Catalase', 'CghF', 'Chal_sti_synt_C', 'Chal_sti_synt_N', 'Chlorophyllase2', 'Chorismate_bind', 'Citrate_synt', 'Clz3', 'Clz7', 'CoA_transf_3', 'Cofilin_ADF', 'Condensation', 'CosA', 'Cp1D', 'CpaE', 'CtvD', 'Cu-oxidase', 'Cu-oxidase_2', 'Cu-oxidase_3', 'Cu_amine_oxid', 'Cupin_2', 'Cys_Met_Meta_PP', 'Cyt-b5', 'Cytochrom_B561', 'DAHP_synth_1', 'DAO', 'DDE_1', 'DDE_3', 'DHDPS', 'DIOX_N', 'DIT1_PvcA', 'DJ-1_PfpI', 'DLH', 'DSBA', 'DUF1295', 'DUF1479', 'DUF1729', 'DUF1749', 'DUF1752', 'DUF1772', 'DUF1776', 'DUF1857', 'DUF1929', 'DUF1965', 'DUF1993', 'DUF2306', 'DUF2334', 'DUF2470', 'DUF2784', 'DUF3176', 'DUF3237', 'DUF3425', 'DUF3471', 'DUF3533', 'DUF4149', 'DUF4185', 'DUF4188', 'DUF4267', 'DUF4385', 'DUF4413', 'DUF4451', 'DUF5519', 'DUF6534', 'DUF6570', 'DUF659', 'DUF87', 'Dabb', 'Dala_Dala_lig_C', 'Dimer_Tnp_hAT', 'Dimerisation', 'Dimerisation2', 'Dynamin_N', 'Dyp_perox', 'ECH_1', 'ECH_2', 'EHN', 'ERG4_ERG24', 'EamA', 'EcKL', 'EcdJ', 'EfHP', 'Egh16-like', 'ElcH', 'Epimerase', 'EqxR', 'Esterase', 'EthD', 'F420_oxidored', 'F5_F8_type_C', 'FAA_hydrolase', 'FAD_binding_1', 'FAD_binding_2', 'FAD_binding_3', 'FAD_binding_4', 'FAD_binding_6', 'FAD_oxidored', 'FAE1_CUT1_RppA', 'FAS_I_H', 'FAS_N', 'FAS_meander', 'FA_hydroxylase', 'FMN_dh', 'FMN_red', 'FMO-like', 'FPSE_09189', 'FR47', 'FSH1', 'Fas_alpha_ACP', 'Fasciclin', 'Fe-ADH', 'Fe-ADH_2', 'Ferric_reduct', 'FhuF', 'Flavodoxin_1', 'Flavodoxin_2', 'Fn3-like', 'FncF', 'FolB', 'FrsA-like', 'FrzJ', 'Fsa2', 'FscB', 'FtsJ', 'FtsK_SpoIIIE', 'Fungal_trans', 'Fungal_trans_2', 'GATase', 'GATase_6', 'GATase_7', 'GBBH-like_N', 'GDP_Man_Dehyd', 'GFO_IDH_MocA', 'GFO_IDH_MocA_C', 'GGPS', 'GH3', 'GH43_C2', 'GIDA', 'GMC_oxred_C', 'GMC_oxred_N', 'GO-like_E_set', 'GST_C', 'GST_C_2', 'GST_C_3', 'GST_C_5', 'GST_N', 'GST_N_2', 'GST_N_3', 'GST_N_4', 'GTP_cyclohydroI', 'G_glu_transpept', 'GliK', 'GliN', 'Gln-synt_C', 'Glu_dehyd_C', 'Gly_transf_sug', 'Glyco_hydro_11', 'Glyco_hydro_3', 'Glyco_hydro_31', 'Glyco_hydro_3_C', 'Glyco_hydro_43', 'Glyco_hydro_cc', 'Glyco_tran_28_C', 'Glyco_transf_28', 'Glyoxal_oxid_N', 'Glyoxalase', 'Glyoxalase_3', 'Glyoxalase_4', 'Glyoxalase_6', 'Gp_dh_C', 'Gp_dh_N', 'Gpr1_Fun34_YaaH', 'GsfR1', 'HAD_2', 'HD', 'HI0933_like', 'HLH', 'HMG-CoA_red', 'HMGL-like', 'HMG_CoA_synt_C', 'HMG_CoA_synt_N', 'HPIH', 'HSP20', 'HTH_51', 'HTH_Tnp_Tc5', 'HTH_psq', 'Haem_degrading', 'HbpS-like', 'Helitron_like_N', 'Helo_like_N', 'Hemerythrin', 'Hemocyanin_C', 'Hemocyanin_M', 'Hexapep', 'HimD', 'Histone', 'Histone_H2A_C', 'Homeobox_KN', 'Homeodomain', 'Homoserine_dh', 'HpcH_HpaI', 'Hydrolase', 'Hydrolase_4', 'Hydrolase_like', 'ICMT', 'IDO', 'ILVD_EDD', 'IMPDH', 'IPPT', 'IPT', 'Iso_dh', 'Isochorismatase', 'IucA_IucC', 'K_oxygenase', 'Kdo', 'Kelch_1', 'Kelch_2', 'Kelch_3', 'Kelch_4', 'Kelch_6', 'LCM', 'LIP', 'Lactamase_B', 'Lactamase_B_2', 'Lactamase_B_3', 'LcsT', 'LigB', 'Lipocalin_5', 'Lipocalin_9', 'LolU', 'LtmS', 'Lumazine_bd_2', 'Lyase_1', 'Lyase_aromatic', 'Lycopene_cycl', 'Lys_Orn_oxgnase', 'Lysine_decarbox', 'M20_dimer', 'MARVEL', 'MBOAT_2', 'MFS_1', 'MFS_1_like', 'MFS_2', 'MFS_4', 'MFS_5', 'MMR_HSR1', 'MPAB_Lcp_cat', 'MTS', 'MULE', 'MafB19-deam', 'MaoC_dehydrat_N', 'MaoC_dehydratas', 'MdpA', 'Melibiase_2', 'Melibiase_C', 'MetW', 'Methyltransf_10', 'Methyltransf_11', 'Methyltransf_12', 'Methyltransf_16', 'Methyltransf_2', 'Methyltransf_21', 'Methyltransf_23', 'Methyltransf_24', 'Methyltransf_25', 'Methyltransf_28', 'Methyltransf_3', 'Methyltransf_31', 'Methyltransf_32', 'Methyltransf_33', 'Methyltransf_4', 'Methyltransf_7', 'Methyltransf_9', 'Mito_carr', 'MmgE_PrpD', 'MmgE_PrpD_C', 'MscL', 'Myb_DNA-bind_6', 'Myb_DNA-bind_7', 'Myb_DNA-binding', 'NACHT', 'NAD_binding_1', 'NAD_binding_10', 'NAD_binding_11', 'NAD_binding_2', 'NAD_binding_3', 'NAD_binding_4', 'NAD_binding_6', 'NAD_binding_8', 'NAD_binding_9', 'NMO', 'NRPS_C', 'NRPS_CT', 'NRPS_E', 'NRPS_term_dom', 'Nitroreductase', 'NmrA', 'NnrU', 'NodS', 'NtCtMGAM_N', 'NvfI', 'O-FucT', 'OATP', 'OPT', 'OphC', 'Orn_Arg_deC_N', 'Orn_DAP_Arg_deC', 'OtaR1', 'Oxidored_FMN', 'Oxysterol_BP', 'P5CR_dimer', 'PAP2_3', 'PBP', 'PCMT', 'PDR_CDR', 'PEMT', 'PEP_mutase', 'PHM7_cyt', 'PIF1', 'PIGO_PIGG', 'PKS_AT', 'PKS_KS', 'PKS_NRPS_MT', 'PK_Tyr_Ser-Thr', 'PPR_2', 'PSDC', 'PS_Dcarbxylase', 'PTR2', 'P_proprotein', 'PatF', 'PatJ', 'Patched', 'PaxC', 'PbcA', 'Pec_lyase', 'PenF', 'Peptidase_M19', 'Peptidase_M20', 'Peptidase_M24', 'Peptidase_M3', 'Peptidase_S10', 'Peptidase_S15', 'Peptidase_S28', 'Peptidase_S41', 'Peptidase_S8', 'Peptidase_S9', 'Peptidase_S9_N', 'Peroxidase_2', 'PfB', 'Phe_hydrox_dim', 'PhomB', 'Phosphodiest', 'PhyH', 'Pkinase', 'Polyketide_cyc2', 'Polysacc_deac_1', 'Polysacc_synt_2', 'Prefoldin', 'Prenyltrans', 'Pribosyltran', 'Pro_CA', 'Pro_isomerase', 'Proteasome', 'PrpF', 'PydY', 'Pyr4', 'Pyr_redox', 'Pyr_redox_2', 'Pyr_redox_3', 'Pyridoxal_deC', 'Questin_oxidase', 'RSN1_7TM', 'RSN1_TM', 'RTA1', 'RmlD_sub_bind', 'RsgA_GTPase', 'Rua1_C', 'SAT2', 'SET', 'SGL', 'SMC_N', 'SQHop_cyclase_C', 'SQHop_cyclase_N', 'SQS_PSY', 'Sacchrp_dh_NADP', 'SalTPS', 'Say1_Mug180', 'Scytalone_dh', 'Sdh_cyt', 'SdnG', 'SdnJ', 'Semialdhyde_dh', 'Shikimate_DH', 'SmbD', 'SnoaL', 'SnoaL_2', 'SnoaL_4', 'Spd14', 'SrdD', 'Steroid_dh', 'Sterol-sensing', 'Sterol_MT_C', 'Str_synth', 'Sugar_tr', 'Sulfotransfer_4', 'T2SSE', 'TFIIE_alpha', 'TIM', 'TPMT', 'TPP_enzyme_C', 'TPP_enzyme_M', 'TPP_enzyme_N', 'TPR_2', 'TP_methylase', 'TRAM1', 'TRAM_LAG1_CLN8', 'TRI12', 'TRI5', 'TRI9', 'TTL', 'Tannase', 'TauD', 'Tcf25', 'TdiC', 'Terpene_syn_C_2', 'Terpene_synth_C', 'Thi4', 'Thioesterase', 'Thiolase_N', 'Tmemb_14', 'TqaK', 'Transferase', 'Transketolase_C', 'Transketolase_N', 'Transp_cyt_pur', 'Tri14', 'Tri3', 'Tri6', 'Trp_DMAT', 'Trp_halogenase', 'Trt14', 'Trt5', 'Trypsin', 'Trypsin_2', 'TwmC', 'Tyrosinase', 'UDPGT', 'UNC-93', 'UbiA_PT', 'UbiA_TC', 'Ubie_methyltran', 'UstYa', 'VicA', 'VicYx', 'VidS', 'VniA', 'VrtF', 'WD40', 'XenC', 'XenD', 'YCII', 'Zn_clus', 'adh_short', 'adh_short_C2', 'bZIP_1', 'dCMP_cyt_deam_1', 'fPKS_DH', 'fPKS_ER', 'fPKS_KR', 'fPKS_PT', 'fPKS_R', 'fPKS_TH', 'fn3_3', 'p450', 'peroxidase', 'tRNA-synt_1c', 'tRNA-synt_1c_C', 'tRNA-synt_2', 'tRNA-synt_2c', 'tRNA-synt_2d', 'tRNA_anti-codon', 'zf-C2H2', 'zf-C6H2', 'zf-H2C2_2', 'zf_CCCH_4']\n",
        "\n",
        "  pfam_widget = widgets.Combobox(\n",
        "      placeholder='Type to search Pfam...',\n",
        "      options=pfam_list,\n",
        "      description='Pfam:',\n",
        "      ensure_option=True,\n",
        "      disabled=False\n",
        "  )\n",
        "\n",
        "  # Confirm button\n",
        "  confirm_button = widgets.Button(description=\"Confirm\")\n",
        "\n",
        "  # Function called when button is clicked\n",
        "  def on_confirm_clicked(b):\n",
        "      global pfam\n",
        "      pfam = pfam_widget.value\n",
        "      clear_output()  # Clear the input and button\n",
        "      print(f\"Pfam has been set to: {pfam}\")\n",
        "\n",
        "  confirm_button.on_click(on_confirm_clicked)\n",
        "\n",
        "  # Display widgets\n",
        "  display(pfam_widget, confirm_button)\n",
        "\n",
        "\n",
        "if mode == \"target\":\n",
        "    # File upload widget\n",
        "    fasta_upload = widgets.FileUpload(\n",
        "        accept='.fasta,.fa',  # Accept FASTA files only\n",
        "        multiple=False,       # Only one file\n",
        "        description=\"Upload FASTA\"\n",
        "    )\n",
        "\n",
        "    # Confirm button\n",
        "    confirm_button = widgets.Button(description=\"Confirm\")\n",
        "\n",
        "    # Function called when button is clicked\n",
        "    def on_confirm_clicked(b):\n",
        "        global fasta_path\n",
        "        if len(fasta_upload.value) == 0:\n",
        "            print(\"Please upload a FASTA file first.\")\n",
        "            return\n",
        "        # Get the uploaded file\n",
        "        uploaded_filename = list(fasta_upload.value.keys())[0]\n",
        "        content = fasta_upload.value[uploaded_filename]['content']\n",
        "        # Save to local path\n",
        "        fasta_path = f\"./{uploaded_filename}\"\n",
        "        with open(fasta_path, \"wb\") as f:\n",
        "            f.write(content)\n",
        "        clear_output()\n",
        "        print(f\"FASTA file has been saved to: {fasta_path}\")\n",
        "\n",
        "    confirm_button.on_click(on_confirm_clicked)\n",
        "\n",
        "    # Display widgets\n",
        "    display(fasta_upload, confirm_button)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 100,
          "referenced_widgets": [
            "4f2e2bfcccb34f349fd924570d33a51c",
            "901da8b871394471ab503815fd78289c",
            "9bebeb1f8edc4bc0ae93a512796af9ed",
            "3f4e500ed9b2498cae10bed9007ed824",
            "260f6d4c62364dd9af6f6ab289e34f37",
            "f4eca612f2084c958ed9a1da368f9706"
          ]
        },
        "id": "J61JbxXGJJtj",
        "outputId": "042933b6-449b-4d34-ed57-cb10a3ba015c"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Pfam has been set to: Pyr4\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@markdown - Upload a directory that contains GenBank file(s) from which BGCs will be extracted and provide the path to the directory. The GenBank files need to have CDS features with a translation qualifier. Do not include any non-GenBank files in the directory.\n",
        "results_dir = \"/content/result\"\n",
        "mode = \"all\" #@param [\"all\", \"target\", \"pfam\", \"ripps\", \"sre\", \"human\"]\n",
        "#@markdown - Choose a mode for BGC extraction.\n",
        "#@markdown  - `all`: Extract all possible BGCs.\n",
        "#@markdown  - `target`: Extract BGCs that encode a homologue of protein(s) of your interest (a FASTA file is required).\n",
        "#@markdown  - `pfam`: Extract BGCs that encode a protein with a protein family/domain (Pfam) of your interest.\n",
        "#@markdown  - `ripps`: Extract BGCs that encode a possible precursor peptide for RiPP biosynthesis.\n",
        "#@markdown  - `sre`: Extract BGCs with a possible self-resistance gene.\n",
        "#@markdown  - `human`: Extract BGCs that encode a homologue of human disease-related proteins.\n",
        "fasta = \"content/YourProteins.fasta\" #@param {type:\"string\"}\n",
        "#@markdown - Required when `target` mode is used. Provide the path to a fasta file containing protein(s) of your interest.\n",
        "pfam = \"Pyr4\" #@param ['2-Hacid_dh', '2-Hacid_dh_C', '2OG-FeII_Oxy', '2OG-FeII_Oxy_2', '2OG-FeII_Oxy_3', '2OG-Fe_Oxy_2', '3Beta_HSD', '3HCDH', '3HCDH_N', '4HBT', '4HBT_2', '4HBT_3', 'AA-adenyl-dom', 'AAA_16', 'AAA_19', 'AAA_21', 'AAA_22', 'AAA_23', 'AAA_25', 'AAA_29', 'AAA_30', 'AAA_33', 'AAT', 'AATase', 'AA_kinase', 'AA_permease', 'AA_permease_2', 'ABA3', 'ABC1', 'ABC2_membrane', 'ABC2_membrane_3', 'ABC2_membrane_7', 'ABC_ATPase', 'ABC_membrane', 'ABC_tran', 'ABC_trans_N', 'ABM', 'ACAS_N', 'ACC_central', 'ACPS', 'ACP_PCP', 'ACP_syn_III_C', 'ACT', 'ACT_7', 'ADC', 'ADH_N', 'ADH_N_2', 'ADH_zinc_N', 'ADH_zinc_N_2', 'ADP_ribosyl_GH', 'ADSL_C', 'AIG2_2', 'APH', 'ATEG_03636', 'ATP-grasp', 'ATP-grasp_4', 'ATP-synt_ab', 'ATP-synt_ab_N', 'ATP11', 'ATPgrasp_N', 'ATPgrasp_Ter', 'AXE1', 'Aa_trans', 'Abhydrolase_1', 'Abhydrolase_2', 'Abhydrolase_3', 'Abhydrolase_6', 'Acetyltransf_1', 'Acetyltransf_3', 'Acetyltransf_7', 'Acetyltransf_8', 'Aconitase', 'Aconitase_C', 'Acyl-CoA_dh_1', 'Acyl-CoA_dh_2', 'Acyl-CoA_dh_M', 'Acyl-CoA_dh_N', 'Acyl_transf_1', 'Acyl_transf_3', 'AflR', 'Aldedh', 'Aldo_ket_red', 'Aldolase_II', 'Aldose_epim', 'Alpha-amylase', 'Amidase', 'Amidohydro_1', 'Amidohydro_2', 'Amidohydro_3', 'Amino_oxidase', 'Aminotran_1_2', 'Aminotran_3', 'Aminotran_4', 'Aminotran_5', 'Ank', 'AnkA', 'AnkG', 'Ank_2', 'Ank_3', 'Ank_4', 'Ank_5', 'Anth_synt_I_N', 'Anthrone_oxy', 'ApbA', 'ApbA_C', 'Arginosuc_synth', 'AsR5', 'AsR6', 'Asn_synthase', 'Asp', 'Asp_Glu_race', 'Asp_aminotransf', 'AstC', 'AtaL', 'AtnE', 'BAAT_C', 'BBE', 'BD-FAE', 'Bac_luciferase', 'Beta-lactamase', 'Beta_elim_lyase', 'BfL6', 'BfL8', 'Biotin_carb_C', 'Biotin_carb_N', 'Biotin_lipoyl', 'CBM-like', 'CBS', 'CMAS', 'CMD', 'COesterase', 'CPSase_L_D2', 'CPSase_L_D3', 'CalD', 'Caps_synth', 'Carboxyl_trans', 'Carn_acyltransf', 'Catalase-rel', 'Catalase', 'CghF', 'Chal_sti_synt_C', 'Chal_sti_synt_N', 'Chlorophyllase2', 'Chorismate_bind', 'Citrate_synt', 'Clz3', 'Clz7', 'CoA_transf_3', 'Cofilin_ADF', 'Condensation', 'CosA', 'Cp1D', 'CpaE', 'CtvD', 'Cu-oxidase', 'Cu-oxidase_2', 'Cu-oxidase_3', 'Cu_amine_oxid', 'Cupin_2', 'Cys_Met_Meta_PP', 'Cyt-b5', 'Cytochrom_B561', 'DAHP_synth_1', 'DAO', 'DDE_1', 'DDE_3', 'DHDPS', 'DIOX_N', 'DIT1_PvcA', 'DJ-1_PfpI', 'DLH', 'DSBA', 'DUF1295', 'DUF1479', 'DUF1729', 'DUF1749', 'DUF1752', 'DUF1772', 'DUF1776', 'DUF1857', 'DUF1929', 'DUF1965', 'DUF1993', 'DUF2306', 'DUF2334', 'DUF2470', 'DUF2784', 'DUF3176', 'DUF3237', 'DUF3425', 'DUF3471', 'DUF3533', 'DUF4149', 'DUF4185', 'DUF4188', 'DUF4267', 'DUF4385', 'DUF4413', 'DUF4451', 'DUF5519', 'DUF6534', 'DUF6570', 'DUF659', 'DUF87', 'Dabb', 'Dala_Dala_lig_C', 'Dimer_Tnp_hAT', 'Dimerisation', 'Dimerisation2', 'Dynamin_N', 'Dyp_perox', 'ECH_1', 'ECH_2', 'EHN', 'ERG4_ERG24', 'EamA', 'EcKL', 'EcdJ', 'EfHP', 'Egh16-like', 'ElcH', 'Epimerase', 'EqxR', 'Esterase', 'EthD', 'F420_oxidored', 'F5_F8_type_C', 'FAA_hydrolase', 'FAD_binding_1', 'FAD_binding_2', 'FAD_binding_3', 'FAD_binding_4', 'FAD_binding_6', 'FAD_oxidored', 'FAE1_CUT1_RppA', 'FAS_I_H', 'FAS_N', 'FAS_meander', 'FA_hydroxylase', 'FMN_dh', 'FMN_red', 'FMO-like', 'FPSE_09189', 'FR47', 'FSH1', 'Fas_alpha_ACP', 'Fasciclin', 'Fe-ADH', 'Fe-ADH_2', 'Ferric_reduct', 'FhuF', 'Flavodoxin_1', 'Flavodoxin_2', 'Fn3-like', 'FncF', 'FolB', 'FrsA-like', 'FrzJ', 'Fsa2', 'FscB', 'FtsJ', 'FtsK_SpoIIIE', 'Fungal_trans', 'Fungal_trans_2', 'GATase', 'GATase_6', 'GATase_7', 'GBBH-like_N', 'GDP_Man_Dehyd', 'GFO_IDH_MocA', 'GFO_IDH_MocA_C', 'GGPS', 'GH3', 'GH43_C2', 'GIDA', 'GMC_oxred_C', 'GMC_oxred_N', 'GO-like_E_set', 'GST_C', 'GST_C_2', 'GST_C_3', 'GST_C_5', 'GST_N', 'GST_N_2', 'GST_N_3', 'GST_N_4', 'GTP_cyclohydroI', 'G_glu_transpept', 'GliK', 'GliN', 'Gln-synt_C', 'Glu_dehyd_C', 'Gly_transf_sug', 'Glyco_hydro_11', 'Glyco_hydro_3', 'Glyco_hydro_31', 'Glyco_hydro_3_C', 'Glyco_hydro_43', 'Glyco_hydro_cc', 'Glyco_tran_28_C', 'Glyco_transf_28', 'Glyoxal_oxid_N', 'Glyoxalase', 'Glyoxalase_3', 'Glyoxalase_4', 'Glyoxalase_6', 'Gp_dh_C', 'Gp_dh_N', 'Gpr1_Fun34_YaaH', 'GsfR1', 'HAD_2', 'HD', 'HI0933_like', 'HLH', 'HMG-CoA_red', 'HMGL-like', 'HMG_CoA_synt_C', 'HMG_CoA_synt_N', 'HPIH', 'HSP20', 'HTH_51', 'HTH_Tnp_Tc5', 'HTH_psq', 'Haem_degrading', 'HbpS-like', 'Helitron_like_N', 'Helo_like_N', 'Hemerythrin', 'Hemocyanin_C', 'Hemocyanin_M', 'Hexapep', 'HimD', 'Histone', 'Histone_H2A_C', 'Homeobox_KN', 'Homeodomain', 'Homoserine_dh', 'HpcH_HpaI', 'Hydrolase', 'Hydrolase_4', 'Hydrolase_like', 'ICMT', 'IDO', 'ILVD_EDD', 'IMPDH', 'IPPT', 'IPT', 'Iso_dh', 'Isochorismatase', 'IucA_IucC', 'K_oxygenase', 'Kdo', 'Kelch_1', 'Kelch_2', 'Kelch_3', 'Kelch_4', 'Kelch_6', 'LCM', 'LIP', 'Lactamase_B', 'Lactamase_B_2', 'Lactamase_B_3', 'LcsT', 'LigB', 'Lipocalin_5', 'Lipocalin_9', 'LolU', 'LtmS', 'Lumazine_bd_2', 'Lyase_1', 'Lyase_aromatic', 'Lycopene_cycl', 'Lys_Orn_oxgnase', 'Lysine_decarbox', 'M20_dimer', 'MARVEL', 'MBOAT_2', 'MFS_1', 'MFS_1_like', 'MFS_2', 'MFS_4', 'MFS_5', 'MMR_HSR1', 'MPAB_Lcp_cat', 'MTS', 'MULE', 'MafB19-deam', 'MaoC_dehydrat_N', 'MaoC_dehydratas', 'MdpA', 'Melibiase_2', 'Melibiase_C', 'MetW', 'Methyltransf_10', 'Methyltransf_11', 'Methyltransf_12', 'Methyltransf_16', 'Methyltransf_2', 'Methyltransf_21', 'Methyltransf_23', 'Methyltransf_24', 'Methyltransf_25', 'Methyltransf_28', 'Methyltransf_3', 'Methyltransf_31', 'Methyltransf_32', 'Methyltransf_33', 'Methyltransf_4', 'Methyltransf_7', 'Methyltransf_9', 'Mito_carr', 'MmgE_PrpD', 'MmgE_PrpD_C', 'MscL', 'Myb_DNA-bind_6', 'Myb_DNA-bind_7', 'Myb_DNA-binding', 'NACHT', 'NAD_binding_1', 'NAD_binding_10', 'NAD_binding_11', 'NAD_binding_2', 'NAD_binding_3', 'NAD_binding_4', 'NAD_binding_6', 'NAD_binding_8', 'NAD_binding_9', 'NMO', 'NRPS_C', 'NRPS_CT', 'NRPS_E', 'NRPS_term_dom', 'Nitroreductase', 'NmrA', 'NnrU', 'NodS', 'NtCtMGAM_N', 'NvfI', 'O-FucT', 'OATP', 'OPT', 'OphC', 'Orn_Arg_deC_N', 'Orn_DAP_Arg_deC', 'OtaR1', 'Oxidored_FMN', 'Oxysterol_BP', 'P5CR_dimer', 'PAP2_3', 'PBP', 'PCMT', 'PDR_CDR', 'PEMT', 'PEP_mutase', 'PHM7_cyt', 'PIF1', 'PIGO_PIGG', 'PKS_AT', 'PKS_KS', 'PKS_NRPS_MT', 'PK_Tyr_Ser-Thr', 'PPR_2', 'PSDC', 'PS_Dcarbxylase', 'PTR2', 'P_proprotein', 'PatF', 'PatJ', 'Patched', 'PaxC', 'PbcA', 'Pec_lyase', 'PenF', 'Peptidase_M19', 'Peptidase_M20', 'Peptidase_M24', 'Peptidase_M3', 'Peptidase_S10', 'Peptidase_S15', 'Peptidase_S28', 'Peptidase_S41', 'Peptidase_S8', 'Peptidase_S9', 'Peptidase_S9_N', 'Peroxidase_2', 'PfB', 'Phe_hydrox_dim', 'PhomB', 'Phosphodiest', 'PhyH', 'Pkinase', 'Polyketide_cyc2', 'Polysacc_deac_1', 'Polysacc_synt_2', 'Prefoldin', 'Prenyltrans', 'Pribosyltran', 'Pro_CA', 'Pro_isomerase', 'Proteasome', 'PrpF', 'PydY', 'Pyr4', 'Pyr_redox', 'Pyr_redox_2', 'Pyr_redox_3', 'Pyridoxal_deC', 'Questin_oxidase', 'RSN1_7TM', 'RSN1_TM', 'RTA1', 'RmlD_sub_bind', 'RsgA_GTPase', 'Rua1_C', 'SAT2', 'SET', 'SGL', 'SMC_N', 'SQHop_cyclase_C', 'SQHop_cyclase_N', 'SQS_PSY', 'Sacchrp_dh_NADP', 'SalTPS', 'Say1_Mug180', 'Scytalone_dh', 'Sdh_cyt', 'SdnG', 'SdnJ', 'Semialdhyde_dh', 'Shikimate_DH', 'SmbD', 'SnoaL', 'SnoaL_2', 'SnoaL_4', 'Spd14', 'SrdD', 'Steroid_dh', 'Sterol-sensing', 'Sterol_MT_C', 'Str_synth', 'Sugar_tr', 'Sulfotransfer_4', 'T2SSE', 'TFIIE_alpha', 'TIM', 'TPMT', 'TPP_enzyme_C', 'TPP_enzyme_M', 'TPP_enzyme_N', 'TPR_2', 'TP_methylase', 'TRAM1', 'TRAM_LAG1_CLN8', 'TRI12', 'TRI5', 'TRI9', 'TTL', 'Tannase', 'TauD', 'Tcf25', 'TdiC', 'Terpene_syn_C_2', 'Terpene_synth_C', 'Thi4', 'Thioesterase', 'Thiolase_N', 'Tmemb_14', 'TqaK', 'Transferase', 'Transketolase_C', 'Transketolase_N', 'Transp_cyt_pur', 'Tri14', 'Tri3', 'Tri6', 'Trp_DMAT', 'Trp_halogenase', 'Trt14', 'Trt5', 'Trypsin', 'Trypsin_2', 'TwmC', 'Tyrosinase', 'UDPGT', 'UNC-93', 'UbiA_PT', 'UbiA_TC', 'Ubie_methyltran', 'UstYa', 'VicA', 'VicYx', 'VidS', 'VniA', 'VrtF', 'WD40', 'XenC', 'XenD', 'YCII', 'Zn_clus', 'adh_short', 'adh_short_C2', 'bZIP_1', 'dCMP_cyt_deam_1', 'fPKS_DH', 'fPKS_ER', 'fPKS_KR', 'fPKS_PT', 'fPKS_R', 'fPKS_TH', 'fn3_3', 'p450', 'peroxidase', 'tRNA-synt_1c', 'tRNA-synt_1c_C', 'tRNA-synt_2', 'tRNA-synt_2c', 'tRNA-synt_2d', 'tRNA_anti-codon', 'zf-C2H2', 'zf-C6H2', 'zf-H2C2_2', 'zf_CCCH_4']\n",
        "#@markdown - Required when `pfam` mode is used. Choose a protein domain of your interest.\n",
        "\n",
        "import ipywidgets as widgets\n",
        "from IPython.display import display\n",
        "\n",
        "# pfam ã®å€™è£œï¼ˆã“ã“ã§ã¯ä¸€éƒ¨ã ã‘ä¾‹ç¤ºã€å®Ÿéš›ã¯é•·ã„ãƒªã‚¹ãƒˆã‚’å…¥ã‚Œã‚‹ï¼‰\n",
        "pfam_list = [\n",
        "    '2-Hacid_dh', '2OG-FeII_Oxy', '3Beta_HSD', 'ABC_ATPase', 'Pyr4',\n",
        "    'p450', 'zf-C2H2', 'zf-C6H2', 'tRNA-synt_1c'\n",
        "]\n",
        "\n",
        "pfam_widget = widgets.Combobox(\n",
        "    placeholder='Type to search Pfam...',\n",
        "    options=pfam_list,\n",
        "    description='Pfam:',\n",
        "    ensure_option=True,\n",
        "    disabled=False\n",
        ")\n",
        "\n",
        "display(pfam_widget)\n",
        "\n",
        "#@markdown **Optional settings**\n",
        "gap_allowed = 2500 #@param {type:\"integer\"}\n",
        "#@markdown - Maximum allowed gap (in bp) between two adjacent genes in a cluster, if no correlation is detected between the two genes (Default: 2500)\n",
        "max_bgc_gap = 15000 #@param {type:\"integer\"}\n",
        "#@markdown - Maximum allowed gap (in bp) between two adjacent genes, if they only have a weak correlation (Default: 15000)\n",
        "min_prot_len = 200 #@param {type:\"integer\"}\n",
        "#@markdown - Minimum protein length considered for biosynthetic protein detection (Default: 200)\n",
        "num_of_genes_checked = 20 #@param {type:\"integer\"}\n",
        "#@markdown - Number of genes around the core/target genes to be checked (Default: 20)\n",
        "min_identity = 50 #@param {type:\"integer\"}\n",
        "#@markdown - Minimum identity (in %) required to be considered as a duplicated protein (Default: 50)"
      ],
      "metadata": {
        "id": "581npsL3MKRR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Install dependencies\n",
        "import os\n",
        "import shutil\n",
        "!sudo apt-get install hmmer diamond-aligner\n",
        "!pip install biopython\n",
        "!pip install natsort\n",
        "!pip install tensorflow\n",
        "if os.path.isdir(\"/content/funbgcex\") == False:\n",
        "  !git clone https://github.com/ydmatsd/funbgcex-colab2.git\n",
        "  shutil.move(\"/content/funbgcex-colab2/funbgcex-colab/funbgcex\",\"/content\")"
      ],
      "metadata": {
        "id": "um2ou0yLLBJ_",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "7d6dba9b-7f59-42fd-a943-011e2c65cec2"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  blends-common libdivsufsort3 med-config menu\n",
            "Suggested packages:\n",
            "  blends-doc hmmer-doc menu-l10n gksu | kde-runtime | ktsuss\n",
            "The following NEW packages will be installed:\n",
            "  blends-common diamond-aligner hmmer libdivsufsort3 med-config menu\n",
            "0 upgraded, 6 newly installed, 0 to remove and 35 not upgraded.\n",
            "Need to get 3,656 kB of archives.\n",
            "After this operation, 16.0 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 menu amd64 2.1.47ubuntu4 [354 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 blends-common all 0.7.4ubuntu1 [15.9 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 diamond-aligner amd64 2.0.14-1 [2,076 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libdivsufsort3 amd64 2.0.1-5 [42.8 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/universe amd64 hmmer amd64 3.3.2+dfsg-1 [1,155 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/universe amd64 med-config all 3.7.1 [11.6 kB]\n",
            "Fetched 3,656 kB in 1s (2,872 kB/s)\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 6.)\n",
            "debconf: falling back to frontend: Readline\n",
            "debconf: unable to initialize frontend: Readline\n",
            "debconf: (This frontend requires a controlling tty.)\n",
            "debconf: falling back to frontend: Teletype\n",
            "dpkg-preconfigure: unable to re-open stdin: \n",
            "Selecting previously unselected package menu.\n",
            "(Reading database ... 126371 files and directories currently installed.)\n",
            "Preparing to unpack .../0-menu_2.1.47ubuntu4_amd64.deb ...\n",
            "Unpacking menu (2.1.47ubuntu4) ...\n",
            "Selecting previously unselected package blends-common.\n",
            "Preparing to unpack .../1-blends-common_0.7.4ubuntu1_all.deb ...\n",
            "Unpacking blends-common (0.7.4ubuntu1) ...\n",
            "Selecting previously unselected package diamond-aligner.\n",
            "Preparing to unpack .../2-diamond-aligner_2.0.14-1_amd64.deb ...\n",
            "Unpacking diamond-aligner (2.0.14-1) ...\n",
            "Selecting previously unselected package libdivsufsort3:amd64.\n",
            "Preparing to unpack .../3-libdivsufsort3_2.0.1-5_amd64.deb ...\n",
            "Unpacking libdivsufsort3:amd64 (2.0.1-5) ...\n",
            "Selecting previously unselected package hmmer.\n",
            "Preparing to unpack .../4-hmmer_3.3.2+dfsg-1_amd64.deb ...\n",
            "Unpacking hmmer (3.3.2+dfsg-1) ...\n",
            "Selecting previously unselected package med-config.\n",
            "Preparing to unpack .../5-med-config_3.7.1_all.deb ...\n",
            "Unpacking med-config (3.7.1) ...\n",
            "Setting up diamond-aligner (2.0.14-1) ...\n",
            "Setting up menu (2.1.47ubuntu4) ...\n",
            "Setting up libdivsufsort3:amd64 (2.0.1-5) ...\n",
            "Setting up blends-common (0.7.4ubuntu1) ...\n",
            "Setting up med-config (3.7.1) ...\n",
            "debconf: unable to initialize frontend: Dialog\n",
            "debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78.)\n",
            "debconf: falling back to frontend: Readline\n",
            "Adding group `med' (GID 107) ...\n",
            "Done.\n",
            "Setting up hmmer (3.3.2+dfsg-1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero_v2.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
            "\n",
            "Processing triggers for menu (2.1.47ubuntu4) ...\n",
            "Collecting biopython\n",
            "  Downloading biopython-1.85-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n",
            "Downloading biopython-1.85-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n",
            "\u001b[2K   \u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m34.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.85\n",
            "Requirement already satisfied: natsort in /usr/local/lib/python3.12/dist-packages (8.4.0)\n",
            "Requirement already satisfied: tensorflow in /usr/local/lib/python3.12/dist-packages (2.19.0)\n",
            "Requirement already satisfied: absl-py>=1.0.0 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (1.4.0)\n",
            "Requirement already satisfied: astunparse>=1.6.0 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (1.6.3)\n",
            "Requirement already satisfied: flatbuffers>=24.3.25 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (25.2.10)\n",
            "Requirement already satisfied: gast!=0.5.0,!=0.5.1,!=0.5.2,>=0.2.1 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (0.6.0)\n",
            "Requirement already satisfied: google-pasta>=0.1.1 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (0.2.0)\n",
            "Requirement already satisfied: libclang>=13.0.0 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (18.1.1)\n",
            "Requirement already satisfied: opt-einsum>=2.3.2 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (3.4.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.12/dist-packages (from tensorflow) (25.0)\n",
            "Requirement already satisfied: protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<6.0.0dev,>=3.20.3 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (5.29.5)\n",
            "Requirement already satisfied: requests<3,>=2.21.0 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (2.32.4)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.12/dist-packages (from tensorflow) (75.2.0)\n",
            "Requirement already satisfied: six>=1.12.0 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (1.17.0)\n",
            "Requirement already satisfied: termcolor>=1.1.0 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (3.1.0)\n",
            "Requirement already satisfied: typing-extensions>=3.6.6 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (4.15.0)\n",
            "Requirement already satisfied: wrapt>=1.11.0 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (1.17.3)\n",
            "Requirement already satisfied: grpcio<2.0,>=1.24.3 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (1.74.0)\n",
            "Requirement already satisfied: tensorboard~=2.19.0 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (2.19.0)\n",
            "Requirement already satisfied: keras>=3.5.0 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (3.10.0)\n",
            "Requirement already satisfied: numpy<2.2.0,>=1.26.0 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (2.0.2)\n",
            "Requirement already satisfied: h5py>=3.11.0 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (3.14.0)\n",
            "Requirement already satisfied: ml-dtypes<1.0.0,>=0.5.1 in /usr/local/lib/python3.12/dist-packages (from tensorflow) (0.5.3)\n",
            "Requirement already satisfied: wheel<1.0,>=0.23.0 in /usr/local/lib/python3.12/dist-packages (from astunparse>=1.6.0->tensorflow) (0.45.1)\n",
            "Requirement already satisfied: rich in /usr/local/lib/python3.12/dist-packages (from keras>=3.5.0->tensorflow) (13.9.4)\n",
            "Requirement already satisfied: namex in /usr/local/lib/python3.12/dist-packages (from keras>=3.5.0->tensorflow) (0.1.0)\n",
            "Requirement already satisfied: optree in /usr/local/lib/python3.12/dist-packages (from keras>=3.5.0->tensorflow) (0.17.0)\n",
            "Requirement already satisfied: charset_normalizer<4,>=2 in /usr/local/lib/python3.12/dist-packages (from requests<3,>=2.21.0->tensorflow) (3.4.3)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.12/dist-packages (from requests<3,>=2.21.0->tensorflow) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.12/dist-packages (from requests<3,>=2.21.0->tensorflow) (2.5.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.12/dist-packages (from requests<3,>=2.21.0->tensorflow) (2025.8.3)\n",
            "Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.12/dist-packages (from tensorboard~=2.19.0->tensorflow) (3.8.2)\n",
            "Requirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in /usr/local/lib/python3.12/dist-packages (from tensorboard~=2.19.0->tensorflow) (0.7.2)\n",
            "Requirement already satisfied: werkzeug>=1.0.1 in /usr/local/lib/python3.12/dist-packages (from tensorboard~=2.19.0->tensorflow) (3.1.3)\n",
            "Requirement already satisfied: MarkupSafe>=2.1.1 in /usr/local/lib/python3.12/dist-packages (from werkzeug>=1.0.1->tensorboard~=2.19.0->tensorflow) (3.0.2)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.12/dist-packages (from rich->keras>=3.5.0->tensorflow) (4.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.12/dist-packages (from rich->keras>=3.5.0->tensorflow) (2.19.2)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.12/dist-packages (from markdown-it-py>=2.2.0->rich->keras>=3.5.0->tensorflow) (0.1.2)\n",
            "Cloning into 'funbgcex-colab2'...\n",
            "remote: Enumerating objects: 84, done.\u001b[K\n",
            "remote: Counting objects: 100% (24/24), done.\u001b[K\n",
            "remote: Compressing objects: 100% (24/24), done.\u001b[K\n",
            "remote: Total 84 (delta 9), reused 0 (delta 0), pack-reused 60 (from 2)\u001b[K\n",
            "Receiving objects: 100% (84/84), 59.11 MiB | 17.26 MiB/s, done.\n",
            "Resolving deltas: 100% (17/17), done.\n",
            "Updating files: 100% (45/45), done.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Extract BGCs\n",
        "import os\n",
        "import sys\n",
        "from funbgcex.BGCeXtractor import BGCeXtractor\n",
        "\n",
        "funbgcex_path = \"/content/funbgcex\"\n",
        "\n",
        "#Accept input files\n",
        "if os.path.isdir(gbk_dir) == False:\n",
        "    sys.exit(\"The input directory does not exist.\")\n",
        "if mode == \"all\":\n",
        "    query = \"none\"\n",
        "elif mode == \"target\":\n",
        "    query = fasta\n",
        "    if os.path.isfile(query) == False:\n",
        "        sys.exit(\"The fasta file does not exist.\")\n",
        "    if query == \"none\":\n",
        "        sys.exit(\"Please provide a fasta file.\")\n",
        "elif mode == \"pfam\":\n",
        "    PfamList = []\n",
        "    allHMM = f\"{funbgcex_path}/data/hmm/fungalSM/fungalSM.hmm\"\n",
        "    with open(allHMM,\"r\") as hmm:\n",
        "        for line in hmm:\n",
        "            if line.startswith(\"NAME\"):\n",
        "                name = line.split(\" \")[-1].replace(\"\\n\",\"\")\n",
        "                PfamList.append(name)\n",
        "    query = pfam\n",
        "    if query not in PfamList:\n",
        "        if query != \"none\":\n",
        "            print(f\"{query} does not exist in the database.\")\n",
        "            ToBeContinued = True\n",
        "            while ToBeContinued:\n",
        "                answer = input(\"Do you want to check the available protein family names? y/n: \")\n",
        "                if answer == \"Y\" or answer == \"y\":\n",
        "                    print(PfamList)\n",
        "                    ToBeContinued = False\n",
        "                    sys.exit()\n",
        "                elif answer == \"N\" or answer == \"n\":\n",
        "                    ToBeContinued = False\n",
        "                else:\n",
        "                    pass\n",
        "            sys.exit()\n",
        "        else:\n",
        "            sys.exit(\"Please provide the name of a protein family to be used for BGC extraction.\")\n",
        "else:\n",
        "    sys.exit(\"Please choose 'all', 'target', or 'pfam' as an extraction mode.\")\n",
        "\n",
        "#Start BGC extraction\n",
        "BGCeXtractor(gbk_dir,results_dir,mode,query,gap_allowed,max_bgc_gap,min_prot_len,num_of_genes_checked,min_identity)"
      ],
      "metadata": {
        "id": "qbYzxjDlyMCk",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 184
        },
        "outputId": "000937aa-d6f7-4232-ecbe-9779e6bef4e2"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "name 'gap_allowed' is not defined",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-857745361.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     47\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     48\u001b[0m \u001b[0;31m#Start BGC extraction\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 49\u001b[0;31m \u001b[0mBGCeXtractor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgbk_dir\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mresults_dir\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mgap_allowed\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmax_bgc_gap\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmin_prot_len\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mnum_of_genes_checked\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmin_identity\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m: name 'gap_allowed' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Download results\n",
        "import shutil\n",
        "from google.colab import files\n",
        "\n",
        "shutil.make_archive(\"/content/results\", 'zip', results_dir)\n",
        "shutil.rmtree(results_dir)\n",
        "files.download(\"/content/results.zip\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "YHHSUKW_2wNb",
        "outputId": "33787622-3a34-499e-c871-ab205f278192",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_0072a26a-1f51-436d-ac5a-daddf6871377\", \"results.zip\", 372273)"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "3TYvM043bVYa"
      }
    }
  ]
}